{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Setup\" data-toc-modified-id=\"Setup-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Setup</a></span><ul class=\"toc-item\"><li><span><a href=\"#Check-your-sample-table-(i.e.-metadata-file)\" data-toc-modified-id=\"Check-your-sample-table-(i.e.-metadata-file)-1.1\"><span class=\"toc-item-num\">1.1&nbsp;&nbsp;</span>Check your sample table (i.e. metadata file)</a></span></li><li><span><a href=\"#Check-your-TRN\" data-toc-modified-id=\"Check-your-TRN-1.2\"><span class=\"toc-item-num\">1.2&nbsp;&nbsp;</span>Check your TRN</a></span></li><li><span><a href=\"#Load-the-data\" data-toc-modified-id=\"Load-the-data-1.3\"><span class=\"toc-item-num\">1.3&nbsp;&nbsp;</span>Load the data</a></span></li></ul></li><li><span><a href=\"#Regulatory-iModulons\" data-toc-modified-id=\"Regulatory-iModulons-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Regulatory iModulons</a></span></li><li><span><a href=\"#Functional-iModulons\" data-toc-modified-id=\"Functional-iModulons-3\"><span class=\"toc-item-num\">3&nbsp;&nbsp;</span>Functional iModulons</a></span><ul class=\"toc-item\"><li><span><a href=\"#GO-Enrichments\" data-toc-modified-id=\"GO-Enrichments-3.1\"><span class=\"toc-item-num\">3.1&nbsp;&nbsp;</span>GO Enrichments</a></span></li><li><span><a href=\"#KEGG-Enrichments\" data-toc-modified-id=\"KEGG-Enrichments-3.2\"><span class=\"toc-item-num\">3.2&nbsp;&nbsp;</span>KEGG Enrichments</a></span><ul class=\"toc-item\"><li><span><a href=\"#Load-KEGG-mapping\" data-toc-modified-id=\"Load-KEGG-mapping-3.2.1\"><span class=\"toc-item-num\">3.2.1&nbsp;&nbsp;</span>Load KEGG mapping</a></span></li><li><span><a href=\"#Perform-enrichment\" data-toc-modified-id=\"Perform-enrichment-3.2.2\"><span class=\"toc-item-num\">3.2.2&nbsp;&nbsp;</span>Perform enrichment</a></span></li><li><span><a href=\"#Convert-KEGG-IDs-to-human-readable-names\" data-toc-modified-id=\"Convert-KEGG-IDs-to-human-readable-names-3.2.3\"><span class=\"toc-item-num\">3.2.3&nbsp;&nbsp;</span>Convert KEGG IDs to human-readable names</a></span></li></ul></li><li><span><a href=\"#Save-files\" data-toc-modified-id=\"Save-files-3.3\"><span class=\"toc-item-num\">3.3&nbsp;&nbsp;</span>Save files</a></span></li></ul></li><li><span><a href=\"#Check-for-single-gene-iModulons\" data-toc-modified-id=\"Check-for-single-gene-iModulons-4\"><span class=\"toc-item-num\">4&nbsp;&nbsp;</span>Check for single gene iModulons</a></span></li><li><span><a href=\"#Old-characterizations\" data-toc-modified-id=\"Old-characterizations-5\"><span class=\"toc-item-num\">5&nbsp;&nbsp;</span>Old characterizations</a></span></li><li><span><a href=\"#Save-iModulon-object\" data-toc-modified-id=\"Save-iModulon-object-6\"><span class=\"toc-item-num\">6&nbsp;&nbsp;</span>Save iModulon object</a></span></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup\n",
    "This IPython notebook will walk through the steps of characterizing iModulons through the semi-automated tools in PyModulon. You will need:\n",
    "\n",
    "* M and A matrices\n",
    "* Expression data (e.g. `log_tpm_norm.csv`)\n",
    "* Gene table and KEGG/GO annotations (Generated in `gene_annotation.ipynb`)\n",
    "* Sample table, with a column for `project` and `condition`\n",
    "* TRN file\n",
    "\n",
    "Optional:\n",
    "* iModulon table (if you already have some characterized iModulons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from os import path\n",
    "import sys\n",
    "sys.path.append('/home/annie/Documents/multi-Ecoli/strains/')\n",
    "from pymodulon.core import IcaData\n",
    "from pymodulon.plotting import *\n",
    "import pandas as pd\n",
    "import re\n",
    "from Bio.KEGG import REST\n",
    "from tqdm.notebook import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Enter the location of your data here\n",
    "data_dir = path.join('..','data','processed_data')\n",
    "\n",
    "# GO and KEGG annotations are in the 'external' folder\n",
    "external_data = path.join('..','data','external')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check your sample table (i.e. metadata file)\n",
    "Your metadata file will probably have a lot of columns, most of which you may not care about. Feel free to save a secondary copy of your metadata file with only columns that seem relevant to you. The two most important columns are:\n",
    "1. `project`\n",
    "2. `condition`\n",
    "\n",
    "Make sure that these columns exist in your metadata file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project</th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Experiment</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRX10864008</th>\n",
       "      <td>photomixotrophic</td>\n",
       "      <td>photoautotroph</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX10864009</th>\n",
       "      <td>photomixotrophic</td>\n",
       "      <td>photoautotroph</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX10864010</th>\n",
       "      <td>photomixotrophic</td>\n",
       "      <td>photoautotroph</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX10864011</th>\n",
       "      <td>photomixotrophic</td>\n",
       "      <td>glucose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX10864012</th>\n",
       "      <td>photomixotrophic</td>\n",
       "      <td>glucose</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      project       condition\n",
       "Experiment                                   \n",
       "SRX10864008  photomixotrophic  photoautotroph\n",
       "SRX10864009  photomixotrophic  photoautotroph\n",
       "SRX10864010  photomixotrophic  photoautotroph\n",
       "SRX10864011  photomixotrophic         glucose\n",
       "SRX10864012  photomixotrophic         glucose"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metadata = pd.read_csv(path.join(data_dir,'metadata.tsv'),index_col=0,sep='\\t')\n",
    "df_metadata[['project','condition']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "print(df_metadata.project.notnull().all())\n",
    "print(df_metadata.condition.notnull().all())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check your TRN"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each row of the TRN file represents a regulatory interaction.  \n",
    "**Your TRN file must have the following columns:**\n",
    "1. `regulator` - Name of regulator (`/` or `+` characters will be converted to `;`)\n",
    "1. `gene_id` - Locus tag of gene being regulated\n",
    "\n",
    "The following columns are optional, but are helpful to have:\n",
    "1. `regulator_id` - Locus tag of regulator\n",
    "1. `gene_name` - Name of gene (can automatically update this using `name2num`)\n",
    "1. `direction` - Direction of regulation ('+' for activation, '-' for repression, '?' or NaN for unknown)\n",
    "1. `evidence` - Evidence of regulation (e.g. ChIP-exo, qRT-PCR, SELEX, Motif search)\n",
    "1. `PMID` - Reference for regulation\n",
    "\n",
    "You may add any other columns that could help you. TRNs may be saved as either CSV or TSV files. See below for an example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>arsR</td>\n",
       "      <td>Synpcc7942_0688</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ccmR</td>\n",
       "      <td>Synpcc7942_1488</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ccmR</td>\n",
       "      <td>Synpcc7942_1489</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ccmR</td>\n",
       "      <td>Synpcc7942_1490</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ccmR</td>\n",
       "      <td>Synpcc7942_1491</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  regulator          gene_id effect\n",
       "0      arsR  Synpcc7942_0688      -\n",
       "1      ccmR  Synpcc7942_1488      -\n",
       "2      ccmR  Synpcc7942_1489      -\n",
       "3      ccmR  Synpcc7942_1490      -\n",
       "4      ccmR  Synpcc7942_1491      -"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_trn = pd.read_csv(path.join(external_data,'TRN.csv'))\n",
    "df_trn.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `regulator` and `gene_id` must be filled in for each row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "print(df_trn.regulator.notnull().all())\n",
    "print(df_trn.gene_id.notnull().all())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load the data\n",
    "You're now ready to load your IcaData object!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:Optimizing iModulon thresholds, may take 2-3 minutes...\n"
     ]
    }
   ],
   "source": [
    "ica_data = IcaData(M = path.join('../data/processed_data/dimensions/130_optimal/','S.csv'),\n",
    "                   A = path.join('../data/processed_data/dimensions/130_optimal/','A.csv'),\n",
    "                   X = path.join(data_dir,'log_tpm_norm.csv'),\n",
    "                   gene_table = path.join('../data/processed_data/','gene_info.csv'),\n",
    "                   sample_table = path.join(data_dir,'metadata.tsv'),\n",
    "                   trn = path.join(external_data,'TRN.csv'),\n",
    "                   optimize_cutoff=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you don't have a TRN (or have a very minimal TRN), use `threshold_method = 'kmeans'`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ica_data = IcaData(M = path.join(data_dir,'M.csv'),\n",
    "#                    A = path.join(data_dir,'A.csv'),\n",
    "#                    X = path.join(data_dir,'log_tpm_norm.csv'),\n",
    "#                    gene_table = path.join(data_dir,'gene_info.csv'),\n",
    "#                    sample_table = path.join(data_dir,'metadata.tsv'),\n",
    "#                    trn = path.join(data_dir,'TRN.csv'),\n",
    "#                    threshold_method = 'kmeans')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Regulatory iModulons\n",
    "Use `compute_trn_enrichment` to automatically check for Regulatory iModulons. The more complete your TRN, the more regulatory iModulons you'll find."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13</td>\n",
       "      <td>ccmR</td>\n",
       "      <td>1.397716e-10</td>\n",
       "      <td>1.397716e-10</td>\n",
       "      <td>0.227273</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.357143</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21</td>\n",
       "      <td>ntcA</td>\n",
       "      <td>1.258998e-25</td>\n",
       "      <td>2.517997e-25</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.705882</td>\n",
       "      <td>12.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>21</td>\n",
       "      <td>ntcB</td>\n",
       "      <td>3.325249e-12</td>\n",
       "      <td>3.325249e-12</td>\n",
       "      <td>0.375000</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>0.461538</td>\n",
       "      <td>6.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>40</td>\n",
       "      <td>ntcA</td>\n",
       "      <td>2.770449e-09</td>\n",
       "      <td>8.311348e-09</td>\n",
       "      <td>0.175000</td>\n",
       "      <td>0.388889</td>\n",
       "      <td>0.241379</td>\n",
       "      <td>7.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>47</td>\n",
       "      <td>ccmR</td>\n",
       "      <td>4.965229e-10</td>\n",
       "      <td>9.930457e-10</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.571429</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>54</td>\n",
       "      <td>sufR</td>\n",
       "      <td>2.985369e-13</td>\n",
       "      <td>2.985369e-13</td>\n",
       "      <td>0.625000</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   imodulon regulator        pvalue        qvalue  precision    recall  \\\n",
       "0        13      ccmR  1.397716e-10  1.397716e-10   0.227273  0.833333   \n",
       "1        21      ntcA  1.258998e-25  2.517997e-25   0.750000  0.666667   \n",
       "2        21      ntcB  3.325249e-12  3.325249e-12   0.375000  0.600000   \n",
       "3        40      ntcA  2.770449e-09  8.311348e-09   0.175000  0.388889   \n",
       "4        47      ccmR  4.965229e-10  9.930457e-10   0.500000  0.666667   \n",
       "5        54      sufR  2.985369e-13  2.985369e-13   0.625000  0.833333   \n",
       "\n",
       "    f1score    TP  regulon_size  imodulon_size  n_regs  \n",
       "0  0.357143   5.0           6.0           22.0     1.0  \n",
       "1  0.705882  12.0          18.0           16.0     1.0  \n",
       "2  0.461538   6.0          10.0           16.0     1.0  \n",
       "3  0.241379   7.0          18.0           40.0     1.0  \n",
       "4  0.571429   4.0           6.0            8.0     1.0  \n",
       "5  0.714286   5.0           6.0            8.0     1.0  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.compute_trn_enrichment()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can also search for AND/OR combinations of regulators using the `max_regs` argument.\n",
    "\n",
    "Regulator enrichments can be directly saved to the `imodulon_table` using the `save` argument. This saves the enrichment with the lowest q-value to the table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13</td>\n",
       "      <td>ccmR</td>\n",
       "      <td>1.397716e-10</td>\n",
       "      <td>1.397716e-10</td>\n",
       "      <td>0.227273</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.357143</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21</td>\n",
       "      <td>ntcA</td>\n",
       "      <td>1.258998e-25</td>\n",
       "      <td>2.517997e-25</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.705882</td>\n",
       "      <td>12.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>21</td>\n",
       "      <td>ntcB</td>\n",
       "      <td>3.325249e-12</td>\n",
       "      <td>3.325249e-12</td>\n",
       "      <td>0.375000</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>0.461538</td>\n",
       "      <td>6.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>40</td>\n",
       "      <td>ntcA</td>\n",
       "      <td>2.770449e-09</td>\n",
       "      <td>8.311348e-09</td>\n",
       "      <td>0.175000</td>\n",
       "      <td>0.388889</td>\n",
       "      <td>0.241379</td>\n",
       "      <td>7.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>47</td>\n",
       "      <td>ccmR</td>\n",
       "      <td>4.965229e-10</td>\n",
       "      <td>9.930457e-10</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.571429</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>54</td>\n",
       "      <td>sufR</td>\n",
       "      <td>2.985369e-13</td>\n",
       "      <td>2.985369e-13</td>\n",
       "      <td>0.625000</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   imodulon regulator        pvalue        qvalue  precision    recall  \\\n",
       "0        13      ccmR  1.397716e-10  1.397716e-10   0.227273  0.833333   \n",
       "1        21      ntcA  1.258998e-25  2.517997e-25   0.750000  0.666667   \n",
       "2        21      ntcB  3.325249e-12  3.325249e-12   0.375000  0.600000   \n",
       "3        40      ntcA  2.770449e-09  8.311348e-09   0.175000  0.388889   \n",
       "4        47      ccmR  4.965229e-10  9.930457e-10   0.500000  0.666667   \n",
       "5        54      sufR  2.985369e-13  2.985369e-13   0.625000  0.833333   \n",
       "\n",
       "    f1score    TP  regulon_size  imodulon_size  n_regs  \n",
       "0  0.357143   5.0           6.0           22.0     1.0  \n",
       "1  0.705882  12.0          18.0           16.0     1.0  \n",
       "2  0.461538   6.0          10.0           16.0     1.0  \n",
       "3  0.241379   7.0          18.0           40.0     1.0  \n",
       "4  0.571429   4.0           6.0            8.0     1.0  \n",
       "5  0.714286   5.0           6.0            8.0     1.0  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# First search for regulator enrichments with 2 regulators\n",
    "ica_data.compute_trn_enrichment(max_regs=2,save=True)\n",
    "\n",
    "# Next, search for regulator enrichments with just one regulator. This will supercede the 2 regulator enrichments.\n",
    "ica_data.compute_trn_enrichment(max_regs=1,save=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The list of regulatory iModulons are shown below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "57 Total iModulons\n",
      "5 Regulatory iModulons\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ccmR</td>\n",
       "      <td>1.397716e-10</td>\n",
       "      <td>1.397716e-10</td>\n",
       "      <td>0.227273</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.357143</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ntcA</td>\n",
       "      <td>1.258998e-25</td>\n",
       "      <td>2.517997e-25</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.705882</td>\n",
       "      <td>12.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>ntcA</td>\n",
       "      <td>2.770449e-09</td>\n",
       "      <td>8.311348e-09</td>\n",
       "      <td>0.175000</td>\n",
       "      <td>0.388889</td>\n",
       "      <td>0.241379</td>\n",
       "      <td>7.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>ccmR</td>\n",
       "      <td>4.965229e-10</td>\n",
       "      <td>9.930457e-10</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.571429</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>sufR</td>\n",
       "      <td>2.985369e-13</td>\n",
       "      <td>2.985369e-13</td>\n",
       "      <td>0.625000</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   regulator        pvalue        qvalue  precision    recall   f1score    TP  \\\n",
       "13      ccmR  1.397716e-10  1.397716e-10   0.227273  0.833333  0.357143   5.0   \n",
       "21      ntcA  1.258998e-25  2.517997e-25   0.750000  0.666667  0.705882  12.0   \n",
       "40      ntcA  2.770449e-09  8.311348e-09   0.175000  0.388889  0.241379   7.0   \n",
       "47      ccmR  4.965229e-10  9.930457e-10   0.500000  0.666667  0.571429   4.0   \n",
       "54      sufR  2.985369e-13  2.985369e-13   0.625000  0.833333  0.714286   5.0   \n",
       "\n",
       "    regulon_size  imodulon_size  n_regs  \n",
       "13           6.0           22.0     1.0  \n",
       "21          18.0           16.0     1.0  \n",
       "40          18.0           40.0     1.0  \n",
       "47           6.0            8.0     1.0  \n",
       "54           6.0            8.0     1.0  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regulatory_imodulons = ica_data.imodulon_table[ica_data.imodulon_table.regulator.notnull()]\n",
    "print(len(ica_data.imodulon_table),'Total iModulons')\n",
    "print(len(regulatory_imodulons),'Regulatory iModulons')\n",
    "regulatory_imodulons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can rename iModulons in this jupyter notebook, or you can save the iModulon table as a CSV and edit it in Excel.\n",
    "\n",
    "If two iModulons have the same regulator (e.g. 'Reg'), they will be named 'Reg-1' and 'Reg-2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:Duplicate iModulon names detected. iModulon 13 will be renamed to ccmR-1\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 21 will be renamed to ntcA-1\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 40 will be renamed to ntcA-2\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 47 will be renamed to ccmR-2\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  regulator  pvalue  qvalue  precision  recall  f1score  TP  regulon_size  \\\n",
       "0       NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "1       NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "2       NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "3       NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "4       NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "\n",
       "   imodulon_size  n_regs  \n",
       "0            NaN     NaN  \n",
       "1            NaN     NaN  \n",
       "2            NaN     NaN  \n",
       "3            NaN     NaN  \n",
       "4            NaN     NaN  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.rename_imodulons(regulatory_imodulons.regulator.to_dict())\n",
    "ica_data.imodulon_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "regulatory_imodulons = ica_data.imodulon_table[ica_data.imodulon_table.regulator.notnull()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functional iModulons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GO annotations and KEGG pathways/modules were generated in the 1_create_the_gene_table.ipynb notebook. Enrichments will be calculated in this notebook, and further curated in the 3_manual_iModulon_curation notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GO Enrichments"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First load the Gene Ontology annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_ontology</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Synpcc7942_1426</td>\n",
       "      <td>cbbL</td>\n",
       "      <td>carboxysome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Synpcc7942_1534</td>\n",
       "      <td>ago</td>\n",
       "      <td>DNA endonuclease activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Synpcc7942_1535</td>\n",
       "      <td>rbcX</td>\n",
       "      <td>cytoplasm|carboxysome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Synpcc7942_1891</td>\n",
       "      <td>labA</td>\n",
       "      <td>circadian rhythm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Synpcc7942_2462</td>\n",
       "      <td>PSK</td>\n",
       "      <td>ATP binding|D-ribulokinase activity</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id gene_name                        gene_ontology\n",
       "0  Synpcc7942_1426      cbbL                          carboxysome\n",
       "1  Synpcc7942_1534       ago            DNA endonuclease activity\n",
       "2  Synpcc7942_1535      rbcX                cytoplasm|carboxysome\n",
       "3  Synpcc7942_1891      labA                     circadian rhythm\n",
       "4  Synpcc7942_2462       PSK  ATP binding|D-ribulokinase activity"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_GO = pd.read_csv(path.join(external_data,'GO_annotations_curated.csv'),index_col=0)\n",
    "DF_GO.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_GO_enrich = ica_data.compute_annotation_enrichment(DF_GO,'gene_ontology')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>gene_ontology</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ccm-1</td>\n",
       "      <td>carboxysome</td>\n",
       "      <td>2.958085e-15</td>\n",
       "      <td>1.271976e-13</td>\n",
       "      <td>0.538462</td>\n",
       "      <td>0.636364</td>\n",
       "      <td>0.583333</td>\n",
       "      <td>7.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ccm-1</td>\n",
       "      <td>carboxysome|structural constituent of carboxys...</td>\n",
       "      <td>2.346911e-08</td>\n",
       "      <td>3.363906e-07</td>\n",
       "      <td>0.307692</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.380952</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ccm-1</td>\n",
       "      <td>structural constituent of carboxysome</td>\n",
       "      <td>2.346911e-08</td>\n",
       "      <td>3.363906e-07</td>\n",
       "      <td>0.307692</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.380952</td>\n",
       "      <td>4.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ccm-1</td>\n",
       "      <td>carbonate dehydratase activity</td>\n",
       "      <td>2.190737e-05</td>\n",
       "      <td>2.355042e-04</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.266667</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ccm-1</td>\n",
       "      <td>carboxysome|carbonate dehydratase activity</td>\n",
       "      <td>4.870738e-03</td>\n",
       "      <td>4.188835e-02</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  imodulon                                      gene_ontology        pvalue  \\\n",
       "0    ccm-1                                        carboxysome  2.958085e-15   \n",
       "1    ccm-1  carboxysome|structural constituent of carboxys...  2.346911e-08   \n",
       "2    ccm-1              structural constituent of carboxysome  2.346911e-08   \n",
       "3    ccm-1                     carbonate dehydratase activity  2.190737e-05   \n",
       "4    ccm-1         carboxysome|carbonate dehydratase activity  4.870738e-03   \n",
       "\n",
       "         qvalue  precision    recall   f1score   TP  target_set_size  \\\n",
       "0  1.271976e-13   0.538462  0.636364  0.583333  7.0             13.0   \n",
       "1  3.363906e-07   0.307692  0.500000  0.380952  4.0              8.0   \n",
       "2  3.363906e-07   0.307692  0.500000  0.380952  4.0             10.0   \n",
       "3  2.355042e-04   0.153846  1.000000  0.266667  2.0              2.0   \n",
       "4  4.188835e-02   0.076923  1.000000  0.142857  1.0              1.0   \n",
       "\n",
       "   imodulon_size  \n",
       "0           13.0  \n",
       "1           13.0  \n",
       "2           13.0  \n",
       "3           13.0  \n",
       "4           13.0  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_GO_enrich.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## KEGG Enrichments"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load KEGG mapping\n",
    "The `kegg_mapping.csv` file contains KEGG orthologies, pathways, modules, and reactions. Only pathways and modules are relevant to iModulon characterization."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['KEGG_pathway' 'KEGG_module' 'KEGG_reaction']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>database</th>\n",
       "      <th>kegg_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Synpcc7942_0001</td>\n",
       "      <td>KEGG_pathway</td>\n",
       "      <td>map00230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Synpcc7942_0001</td>\n",
       "      <td>KEGG_pathway</td>\n",
       "      <td>map00240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Synpcc7942_0001</td>\n",
       "      <td>KEGG_pathway</td>\n",
       "      <td>map01100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Synpcc7942_0001</td>\n",
       "      <td>KEGG_pathway</td>\n",
       "      <td>map03030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Synpcc7942_0001</td>\n",
       "      <td>KEGG_pathway</td>\n",
       "      <td>map03430</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene_id      database   kegg_id\n",
       "6   Synpcc7942_0001  KEGG_pathway  map00230\n",
       "7   Synpcc7942_0001  KEGG_pathway  map00240\n",
       "8   Synpcc7942_0001  KEGG_pathway  map01100\n",
       "9   Synpcc7942_0001  KEGG_pathway  map03030\n",
       "10  Synpcc7942_0001  KEGG_pathway  map03430"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_KEGG = pd.read_csv(path.join(external_data,'kegg_mapping.csv'),index_col=0)\n",
    "print(DF_KEGG.database.unique())\n",
    "DF_KEGG.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "kegg_pathways = DF_KEGG[DF_KEGG.database == 'KEGG_pathway']\n",
    "kegg_modules = DF_KEGG[DF_KEGG.database == 'KEGG_module']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Perform enrichment\n",
    "Uses the `compute_annotation_enrichment` function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_pathway_enrich = ica_data.compute_annotation_enrichment(kegg_pathways,'kegg_id')\n",
    "DF_module_enrich = ica_data.compute_annotation_enrichment(kegg_modules,'kegg_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PSII</td>\n",
       "      <td>map00195</td>\n",
       "      <td>5.830475e-05</td>\n",
       "      <td>1.195247e-02</td>\n",
       "      <td>0.238095</td>\n",
       "      <td>0.087719</td>\n",
       "      <td>0.128205</td>\n",
       "      <td>5.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ribosome-1</td>\n",
       "      <td>map03010</td>\n",
       "      <td>6.970523e-39</td>\n",
       "      <td>1.428957e-36</td>\n",
       "      <td>0.884615</td>\n",
       "      <td>0.433962</td>\n",
       "      <td>0.582278</td>\n",
       "      <td>23.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UV</td>\n",
       "      <td>map00196</td>\n",
       "      <td>2.293707e-16</td>\n",
       "      <td>4.702100e-14</td>\n",
       "      <td>0.155844</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.258065</td>\n",
       "      <td>12.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>77.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UV</td>\n",
       "      <td>map00195</td>\n",
       "      <td>1.672395e-11</td>\n",
       "      <td>1.714204e-09</td>\n",
       "      <td>0.194805</td>\n",
       "      <td>0.263158</td>\n",
       "      <td>0.223881</td>\n",
       "      <td>15.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>77.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UV</td>\n",
       "      <td>map01100</td>\n",
       "      <td>3.859217e-08</td>\n",
       "      <td>2.637131e-06</td>\n",
       "      <td>0.480519</td>\n",
       "      <td>0.067518</td>\n",
       "      <td>0.118400</td>\n",
       "      <td>37.0</td>\n",
       "      <td>548.0</td>\n",
       "      <td>77.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     imodulon   kegg_id        pvalue        qvalue  precision    recall  \\\n",
       "0        PSII  map00195  5.830475e-05  1.195247e-02   0.238095  0.087719   \n",
       "1  ribosome-1  map03010  6.970523e-39  1.428957e-36   0.884615  0.433962   \n",
       "2          UV  map00196  2.293707e-16  4.702100e-14   0.155844  0.750000   \n",
       "3          UV  map00195  1.672395e-11  1.714204e-09   0.194805  0.263158   \n",
       "4          UV  map01100  3.859217e-08  2.637131e-06   0.480519  0.067518   \n",
       "\n",
       "    f1score    TP  target_set_size  imodulon_size  \n",
       "0  0.128205   5.0             57.0           21.0  \n",
       "1  0.582278  23.0             53.0           26.0  \n",
       "2  0.258065  12.0             16.0           77.0  \n",
       "3  0.223881  15.0             57.0           77.0  \n",
       "4  0.118400  37.0            548.0           77.0  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_pathway_enrich.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ccm-1</td>\n",
       "      <td>M00166</td>\n",
       "      <td>4.537653e-04</td>\n",
       "      <td>7.890851e-02</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ccm-1</td>\n",
       "      <td>M00532</td>\n",
       "      <td>7.736129e-04</td>\n",
       "      <td>7.890851e-02</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>0.181818</td>\n",
       "      <td>2.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ccm-1</td>\n",
       "      <td>M00165</td>\n",
       "      <td>1.406629e-03</td>\n",
       "      <td>9.565078e-02</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.160000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ribosome-1</td>\n",
       "      <td>M00178</td>\n",
       "      <td>6.970523e-39</td>\n",
       "      <td>1.421987e-36</td>\n",
       "      <td>0.884615</td>\n",
       "      <td>0.433962</td>\n",
       "      <td>0.582278</td>\n",
       "      <td>23.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ribosome-1</td>\n",
       "      <td>M00179</td>\n",
       "      <td>3.596169e-35</td>\n",
       "      <td>3.668093e-33</td>\n",
       "      <td>0.730769</td>\n",
       "      <td>0.633333</td>\n",
       "      <td>0.678571</td>\n",
       "      <td>19.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     imodulon kegg_id        pvalue        qvalue  precision    recall  \\\n",
       "0       ccm-1  M00166  4.537653e-04  7.890851e-02   0.153846  0.285714   \n",
       "1       ccm-1  M00532  7.736129e-04  7.890851e-02   0.153846  0.222222   \n",
       "2       ccm-1  M00165  1.406629e-03  9.565078e-02   0.153846  0.166667   \n",
       "3  ribosome-1  M00178  6.970523e-39  1.421987e-36   0.884615  0.433962   \n",
       "4  ribosome-1  M00179  3.596169e-35  3.668093e-33   0.730769  0.633333   \n",
       "\n",
       "    f1score    TP  target_set_size  imodulon_size  \n",
       "0  0.200000   2.0              7.0           13.0  \n",
       "1  0.181818   2.0              9.0           13.0  \n",
       "2  0.160000   2.0             12.0           13.0  \n",
       "3  0.582278  23.0             53.0           26.0  \n",
       "4  0.678571  19.0             30.0           26.0  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_module_enrich.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert KEGG IDs to human-readable names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "df79a7e86dec4a91804b117e5b7b0864",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/51 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8cdbd58f7df846139074173b828e8f55",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/26 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for idx,key in tqdm(DF_pathway_enrich.kegg_id.items(),total=len(DF_pathway_enrich)):\n",
    "    text = REST.kegg_find('pathway',key).read()\n",
    "    try:\n",
    "        name = re.search('\\t(.*)\\n',text).group(1)\n",
    "        DF_pathway_enrich.loc[idx,'pathway_name'] = name\n",
    "    except AttributeError:\n",
    "        DF_pathway_enrich.loc[idx,'pathway_name'] = None\n",
    "    \n",
    "for idx,key in tqdm(DF_module_enrich.kegg_id.items(),total=len(DF_module_enrich)):\n",
    "    text = REST.kegg_find('module',key).read()\n",
    "    try:\n",
    "        name = re.search('\\t(.*)\\n',text).group(1)\n",
    "        DF_module_enrich.loc[idx,'module_name'] = name\n",
    "    except AttributeError:\n",
    "        DF_module_enrich.loc[idx,'module_name'] = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>pathway_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PSII</td>\n",
       "      <td>map00195</td>\n",
       "      <td>5.830475e-05</td>\n",
       "      <td>1.195247e-02</td>\n",
       "      <td>0.238095</td>\n",
       "      <td>0.087719</td>\n",
       "      <td>0.128205</td>\n",
       "      <td>5.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>Photosynthesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ribosome-1</td>\n",
       "      <td>map03010</td>\n",
       "      <td>6.970523e-39</td>\n",
       "      <td>1.428957e-36</td>\n",
       "      <td>0.884615</td>\n",
       "      <td>0.433962</td>\n",
       "      <td>0.582278</td>\n",
       "      <td>23.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>Ribosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UV</td>\n",
       "      <td>map00196</td>\n",
       "      <td>2.293707e-16</td>\n",
       "      <td>4.702100e-14</td>\n",
       "      <td>0.155844</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.258065</td>\n",
       "      <td>12.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>Photosynthesis - antenna proteins</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UV</td>\n",
       "      <td>map00195</td>\n",
       "      <td>1.672395e-11</td>\n",
       "      <td>1.714204e-09</td>\n",
       "      <td>0.194805</td>\n",
       "      <td>0.263158</td>\n",
       "      <td>0.223881</td>\n",
       "      <td>15.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>Photosynthesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UV</td>\n",
       "      <td>map01100</td>\n",
       "      <td>3.859217e-08</td>\n",
       "      <td>2.637131e-06</td>\n",
       "      <td>0.480519</td>\n",
       "      <td>0.067518</td>\n",
       "      <td>0.118400</td>\n",
       "      <td>37.0</td>\n",
       "      <td>548.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>Metabolic pathways</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     imodulon   kegg_id        pvalue        qvalue  precision    recall  \\\n",
       "0        PSII  map00195  5.830475e-05  1.195247e-02   0.238095  0.087719   \n",
       "1  ribosome-1  map03010  6.970523e-39  1.428957e-36   0.884615  0.433962   \n",
       "2          UV  map00196  2.293707e-16  4.702100e-14   0.155844  0.750000   \n",
       "3          UV  map00195  1.672395e-11  1.714204e-09   0.194805  0.263158   \n",
       "4          UV  map01100  3.859217e-08  2.637131e-06   0.480519  0.067518   \n",
       "\n",
       "    f1score    TP  target_set_size  imodulon_size  \\\n",
       "0  0.128205   5.0             57.0           21.0   \n",
       "1  0.582278  23.0             53.0           26.0   \n",
       "2  0.258065  12.0             16.0           77.0   \n",
       "3  0.223881  15.0             57.0           77.0   \n",
       "4  0.118400  37.0            548.0           77.0   \n",
       "\n",
       "                        pathway_name  \n",
       "0                     Photosynthesis  \n",
       "1                           Ribosome  \n",
       "2  Photosynthesis - antenna proteins  \n",
       "3                     Photosynthesis  \n",
       "4                 Metabolic pathways  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_pathway_enrich.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>module_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ccm-1</td>\n",
       "      <td>M00166</td>\n",
       "      <td>4.537653e-04</td>\n",
       "      <td>7.890851e-02</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ccm-1</td>\n",
       "      <td>M00532</td>\n",
       "      <td>7.736129e-04</td>\n",
       "      <td>7.890851e-02</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>0.181818</td>\n",
       "      <td>2.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>Photorespiration</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ccm-1</td>\n",
       "      <td>M00165</td>\n",
       "      <td>1.406629e-03</td>\n",
       "      <td>9.565078e-02</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.160000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>Reductive pentose phosphate cycle (Calvin cycle)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ribosome-1</td>\n",
       "      <td>M00178</td>\n",
       "      <td>6.970523e-39</td>\n",
       "      <td>1.421987e-36</td>\n",
       "      <td>0.884615</td>\n",
       "      <td>0.433962</td>\n",
       "      <td>0.582278</td>\n",
       "      <td>23.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ribosome-1</td>\n",
       "      <td>M00179</td>\n",
       "      <td>3.596169e-35</td>\n",
       "      <td>3.668093e-33</td>\n",
       "      <td>0.730769</td>\n",
       "      <td>0.633333</td>\n",
       "      <td>0.678571</td>\n",
       "      <td>19.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     imodulon kegg_id        pvalue        qvalue  precision    recall  \\\n",
       "0       ccm-1  M00166  4.537653e-04  7.890851e-02   0.153846  0.285714   \n",
       "1       ccm-1  M00532  7.736129e-04  7.890851e-02   0.153846  0.222222   \n",
       "2       ccm-1  M00165  1.406629e-03  9.565078e-02   0.153846  0.166667   \n",
       "3  ribosome-1  M00178  6.970523e-39  1.421987e-36   0.884615  0.433962   \n",
       "4  ribosome-1  M00179  3.596169e-35  3.668093e-33   0.730769  0.633333   \n",
       "\n",
       "    f1score    TP  target_set_size  imodulon_size  \\\n",
       "0  0.200000   2.0              7.0           13.0   \n",
       "1  0.181818   2.0              9.0           13.0   \n",
       "2  0.160000   2.0             12.0           13.0   \n",
       "3  0.582278  23.0             53.0           26.0   \n",
       "4  0.678571  19.0             30.0           26.0   \n",
       "\n",
       "                                        module_name  \n",
       "0                                               NaN  \n",
       "1                                  Photorespiration  \n",
       "2  Reductive pentose phosphate cycle (Calvin cycle)  \n",
       "3                                              None  \n",
       "4                                              None  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_module_enrich.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_GO_enrich['source'] = 'GO'\n",
    "DF_pathway_enrich['source'] = 'KEGG pathways'\n",
    "DF_module_enrich['source'] = 'KEGG modules'\n",
    "\n",
    "DF_GO_enrich.rename({'gene_ontology':'annotation'},axis=1, inplace=True)\n",
    "DF_pathway_enrich.rename({'kegg_id':'annotation'},axis=1, inplace=True)\n",
    "DF_module_enrich.rename({'kegg_id':'annotation'},axis=1, inplace=True)\n",
    "\n",
    "DF_enrichments = pd.concat([DF_GO_enrich, DF_pathway_enrich, DF_module_enrich])\n",
    "DF_enrichments.to_csv(path.join(data_dir,'functional_enrichments.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check for single gene iModulons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some iModulons are dominated by a single, high-coefficient gene. These iModulons may result from:\n",
    "1. Overdecomposition of the dataset to identify noisy genes\n",
    "1. Artificial knock-out of single genes\n",
    "1. Regulons with only one target gene\n",
    "\n",
    "No matter what causes these iModulons, it is important to be aware of them. The find_single_gene_imodulons function identifies iModulons that are likely dominated by a single gene.\n",
    "\n",
    "The iModulons identified by ``find_single_gene_imodulons`` may contain more than one gene, since a threshold-agnostic method is used to identify these iModulons."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg_imods = ica_data.find_single_gene_imodulons(save=True)\n",
    "len(sg_imods)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i,mod in enumerate(sg_imods):\n",
    "    ica_data.rename_imodulons({mod:'SG_'+str(i+1)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "      <th>single_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SG_1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_6</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_7</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_8</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_9</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_10</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_11</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_12</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_13</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_14</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      regulator  pvalue  qvalue  precision  recall  f1score  TP  regulon_size  \\\n",
       "SG_1        NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "SG_2        NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "SG_3        NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "SG_4        NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "SG_5        NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "SG_6        NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "SG_7        NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "SG_8        NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "SG_9        NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "SG_10       NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "SG_11       NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "SG_12       NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "SG_13       NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "SG_14       NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "\n",
       "       imodulon_size  n_regs single_gene  \n",
       "SG_1             NaN     NaN        True  \n",
       "SG_2             NaN     NaN        True  \n",
       "SG_3             NaN     NaN        True  \n",
       "SG_4             NaN     NaN        True  \n",
       "SG_5             NaN     NaN        True  \n",
       "SG_6             NaN     NaN        True  \n",
       "SG_7             NaN     NaN        True  \n",
       "SG_8             NaN     NaN        True  \n",
       "SG_9             NaN     NaN        True  \n",
       "SG_10            NaN     NaN        True  \n",
       "SG_11            NaN     NaN        True  \n",
       "SG_12            NaN     NaN        True  \n",
       "SG_13            NaN     NaN        True  \n",
       "SG_14            NaN     NaN        True  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.imodulon_table[ica_data.imodulon_table.single_gene == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_weight</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>accession</th>\n",
       "      <th>old_locus_tag</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_product</th>\n",
       "      <th>COG</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>operon</th>\n",
       "      <th>regulator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Synpcc7942_0601</th>\n",
       "      <td>0.095330</td>\n",
       "      <td>Synpcc7942_0601</td>\n",
       "      <td>CP000100.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>590130</td>\n",
       "      <td>590390</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>No COG annotation</td>\n",
       "      <td>Q31QN6</td>\n",
       "      <td>Op469</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Synpcc7942_0726</th>\n",
       "      <td>0.074952</td>\n",
       "      <td>Synpcc7942_0726</td>\n",
       "      <td>CP000100.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>720068</td>\n",
       "      <td>720430</td>\n",
       "      <td>+</td>\n",
       "      <td>conserved hypothetical protein</td>\n",
       "      <td>No COG annotation</td>\n",
       "      <td>Q31QB1</td>\n",
       "      <td>Op565</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Synpcc7942_0845</th>\n",
       "      <td>0.643671</td>\n",
       "      <td>Synpcc7942_0845</td>\n",
       "      <td>CP000100.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>839873</td>\n",
       "      <td>840079</td>\n",
       "      <td>+</td>\n",
       "      <td>conserved hypothetical protein</td>\n",
       "      <td>No COG annotation</td>\n",
       "      <td>Q31PZ3</td>\n",
       "      <td>Op648</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Synpcc7942_1041</th>\n",
       "      <td>0.079109</td>\n",
       "      <td>Synpcc7942_1041</td>\n",
       "      <td>CP000100.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1055349</td>\n",
       "      <td>1055621</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>No COG annotation</td>\n",
       "      <td>Q31PE8</td>\n",
       "      <td>Op795</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Synpcc7942_2065</th>\n",
       "      <td>0.090690</td>\n",
       "      <td>Synpcc7942_2065</td>\n",
       "      <td>CP000100.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2141928</td>\n",
       "      <td>2142182</td>\n",
       "      <td>-</td>\n",
       "      <td>conserved hypothetical protein</td>\n",
       "      <td>No COG annotation</td>\n",
       "      <td>Q31LH4</td>\n",
       "      <td>Op1555</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Synpcc7942_2233</th>\n",
       "      <td>0.086502</td>\n",
       "      <td>Synpcc7942_2233</td>\n",
       "      <td>CP000100.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2299796</td>\n",
       "      <td>2300113</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>No COG annotation</td>\n",
       "      <td>Q31L06</td>\n",
       "      <td>Op1674</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gene_weight        gene_name   accession old_locus_tag  \\\n",
       "Synpcc7942_0601     0.095330  Synpcc7942_0601  CP000100.1           NaN   \n",
       "Synpcc7942_0726     0.074952  Synpcc7942_0726  CP000100.1           NaN   \n",
       "Synpcc7942_0845     0.643671  Synpcc7942_0845  CP000100.1           NaN   \n",
       "Synpcc7942_1041     0.079109  Synpcc7942_1041  CP000100.1           NaN   \n",
       "Synpcc7942_2065     0.090690  Synpcc7942_2065  CP000100.1           NaN   \n",
       "Synpcc7942_2233     0.086502  Synpcc7942_2233  CP000100.1           NaN   \n",
       "\n",
       "                   start      end strand                    gene_product  \\\n",
       "Synpcc7942_0601   590130   590390      -            hypothetical protein   \n",
       "Synpcc7942_0726   720068   720430      +  conserved hypothetical protein   \n",
       "Synpcc7942_0845   839873   840079      +  conserved hypothetical protein   \n",
       "Synpcc7942_1041  1055349  1055621      -            hypothetical protein   \n",
       "Synpcc7942_2065  2141928  2142182      -  conserved hypothetical protein   \n",
       "Synpcc7942_2233  2299796  2300113      -            hypothetical protein   \n",
       "\n",
       "                               COG uniprot  operon regulator  \n",
       "Synpcc7942_0601  No COG annotation  Q31QN6   Op469       NaN  \n",
       "Synpcc7942_0726  No COG annotation  Q31QB1   Op565       NaN  \n",
       "Synpcc7942_0845  No COG annotation  Q31PZ3   Op648       NaN  \n",
       "Synpcc7942_1041  No COG annotation  Q31PE8   Op795       NaN  \n",
       "Synpcc7942_2065  No COG annotation  Q31LH4  Op1555       NaN  \n",
       "Synpcc7942_2233  No COG annotation  Q31L06  Op1674       NaN  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.view_imodulon('SG_1')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Old characterizations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon.io import *\n",
    "from pymodulon.compare import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "old = load_json_model('../Tahani/selon.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>subject</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Synpcc7942_R0020</td>\n",
       "      <td>SYNPCC7942_RS12910</td>\n",
       "      <td>tRNA-Asp1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Synpcc7942_R0033</td>\n",
       "      <td>SYNPCC7942_RS02990</td>\n",
       "      <td>tRNA-Ile1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Synpcc7942_R0016</td>\n",
       "      <td>SYNPCC7942_RS12495</td>\n",
       "      <td>tRNA-Ser2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Synpcc7942_R0048</td>\n",
       "      <td>SYNPCC7942_RS07055</td>\n",
       "      <td>tRNA-Leu3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Synpcc7942_R0030</td>\n",
       "      <td>SYNPCC7942_RS01480</td>\n",
       "      <td>tRNA-Ala3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene             subject       name\n",
       "0  Synpcc7942_R0020  SYNPCC7942_RS12910  tRNA-Asp1\n",
       "1  Synpcc7942_R0033  SYNPCC7942_RS02990  tRNA-Ile1\n",
       "2  Synpcc7942_R0016  SYNPCC7942_RS12495  tRNA-Ser2\n",
       "3  Synpcc7942_R0048  SYNPCC7942_RS07055  tRNA-Leu3\n",
       "4  Synpcc7942_R0030  SYNPCC7942_RS01480  tRNA-Ala3"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping = pd.read_csv('../data/external/1_gene_mapping_biocyc.csv')\n",
    "mapping = mapping.rename(columns={'objecct_id':'gene','new_locus':'subject'})\n",
    "mapping['gene'] = mapping['gene'].str.replace('SYNPCC7942', 'Synpcc7942')\n",
    "mapping.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.43.0 (0)\n",
       " -->\n",
       "<!-- Title: %3 Pages: 1 -->\n",
       "<svg width=\"720pt\" height=\"531pt\"\n",
       " viewBox=\"0.00 0.00 720.00 530.62\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1.17 1.17) rotate(0) translate(4 618)\">\n",
       "<title>%3</title>\n",
       "<polygon fill=\"white\" stroke=\"transparent\" points=\"-4,4 -4,-618 840,-618 840,4 -4,4\"/>\n",
       "<!-- data2_Biofilm&#45;1 -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>data2_Biofilm&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"39.5\" y=\"-533.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Biofilm&#45;1</text>\n",
       "</g>\n",
       "<!-- data2_Biofilm&#45;2 -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>data2_Biofilm&#45;2</title>\n",
       "<text text-anchor=\"middle\" x=\"213.5\" y=\"-533.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Biofilm&#45;2</text>\n",
       "</g>\n",
       "<!-- data2_CMM -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>data2_CMM</title>\n",
       "<text text-anchor=\"middle\" x=\"373.5\" y=\"-533.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">CMM</text>\n",
       "</g>\n",
       "<!-- data2_Carbon Transport -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>data2_Carbon Transport</title>\n",
       "<text text-anchor=\"middle\" x=\"505.5\" y=\"-540.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Carbon T&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"505.5\" y=\"-525.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ransport</text>\n",
       "</g>\n",
       "<!-- data2_Cell Division -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>data2_Cell Division</title>\n",
       "<text text-anchor=\"middle\" x=\"647.5\" y=\"-540.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Cell D&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"647.5\" y=\"-525.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ivision</text>\n",
       "</g>\n",
       "<!-- data2_CmpR -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>data2_CmpR</title>\n",
       "<text text-anchor=\"middle\" x=\"578.5\" y=\"-533.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">CmpR</text>\n",
       "</g>\n",
       "<!-- data2_Comp&#45;2 -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>data2_Comp&#45;2</title>\n",
       "<text text-anchor=\"middle\" x=\"788.5\" y=\"-533.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Comp&#45;2</text>\n",
       "</g>\n",
       "<!-- data2_CysR -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>data2_CysR</title>\n",
       "<text text-anchor=\"middle\" x=\"67.5\" y=\"-429.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">CysR</text>\n",
       "</g>\n",
       "<!-- data2_Cytc Oxidases  -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>data2_Cytc Oxidases </title>\n",
       "<text text-anchor=\"middle\" x=\"213.5\" y=\"-436.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Cytc Ox&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"213.5\" y=\"-421.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">idases </text>\n",
       "</g>\n",
       "<!-- data2_Dark&#45;Induced Transition -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>data2_Dark&#45;Induced Transition</title>\n",
       "<text text-anchor=\"middle\" x=\"373.5\" y=\"-436.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Dark&#45;Induce&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"373.5\" y=\"-421.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">d Transition</text>\n",
       "</g>\n",
       "<!-- data2_Fur&#45;Like -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>data2_Fur&#45;Like</title>\n",
       "<text text-anchor=\"middle\" x=\"535.5\" y=\"-429.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Fur&#45;Like</text>\n",
       "</g>\n",
       "<!-- data2_HSR -->\n",
       "<g id=\"node12\" class=\"node\">\n",
       "<title>data2_HSR</title>\n",
       "<text text-anchor=\"middle\" x=\"674.5\" y=\"-429.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">HSR</text>\n",
       "</g>\n",
       "<!-- data2_IdiB -->\n",
       "<g id=\"node13\" class=\"node\">\n",
       "<title>data2_IdiB</title>\n",
       "<text text-anchor=\"middle\" x=\"788.5\" y=\"-429.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">IdiB</text>\n",
       "</g>\n",
       "<!-- data2_Membrane Biogenesis -->\n",
       "<g id=\"node14\" class=\"node\">\n",
       "<title>data2_Membrane Biogenesis</title>\n",
       "<text text-anchor=\"middle\" x=\"67.5\" y=\"-332.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Membrane &#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"67.5\" y=\"-317.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Biogenesis</text>\n",
       "</g>\n",
       "<!-- data2_NblS&#45;RpaB&#45;SrrA&#45;2  -->\n",
       "<g id=\"node15\" class=\"node\">\n",
       "<title>data2_NblS&#45;RpaB&#45;SrrA&#45;2 </title>\n",
       "<text text-anchor=\"middle\" x=\"185.5\" y=\"-332.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">NblS&#45;Rpa&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"185.5\" y=\"-317.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">B&#45;SrrA&#45;2 </text>\n",
       "</g>\n",
       "<!-- data2_NtcA&#45;1 -->\n",
       "<g id=\"node16\" class=\"node\">\n",
       "<title>data2_NtcA&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"373.5\" y=\"-325.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">NtcA&#45;1</text>\n",
       "</g>\n",
       "<!-- data2_NtcA&#45;2 -->\n",
       "<g id=\"node17\" class=\"node\">\n",
       "<title>data2_NtcA&#45;2</title>\n",
       "<text text-anchor=\"middle\" x=\"535.5\" y=\"-325.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">NtcA&#45;2</text>\n",
       "</g>\n",
       "<!-- data2_NtcA&#45;3 -->\n",
       "<g id=\"node18\" class=\"node\">\n",
       "<title>data2_NtcA&#45;3</title>\n",
       "<text text-anchor=\"middle\" x=\"674.5\" y=\"-325.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">NtcA&#45;3</text>\n",
       "</g>\n",
       "<!-- data2_NtcA&#45;4 -->\n",
       "<g id=\"node19\" class=\"node\">\n",
       "<title>data2_NtcA&#45;4</title>\n",
       "<text text-anchor=\"middle\" x=\"788.5\" y=\"-325.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">NtcA&#45;4</text>\n",
       "</g>\n",
       "<!-- data2_OxPh NADPH ** -->\n",
       "<g id=\"node20\" class=\"node\">\n",
       "<title>data2_OxPh NADPH **</title>\n",
       "<text text-anchor=\"middle\" x=\"67.5\" y=\"-228.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">OxPh N&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"67.5\" y=\"-213.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ADPH **</text>\n",
       "</g>\n",
       "<!-- data2_OxiTol -->\n",
       "<g id=\"node21\" class=\"node\">\n",
       "<title>data2_OxiTol</title>\n",
       "<text text-anchor=\"middle\" x=\"213.5\" y=\"-221.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">OxiTol</text>\n",
       "</g>\n",
       "<!-- data2_PS -->\n",
       "<g id=\"node22\" class=\"node\">\n",
       "<title>data2_PS</title>\n",
       "<text text-anchor=\"middle\" x=\"256.5\" y=\"-325.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">PS</text>\n",
       "</g>\n",
       "<!-- data2_Phage Related -->\n",
       "<g id=\"node23\" class=\"node\">\n",
       "<title>data2_Phage Related</title>\n",
       "<text text-anchor=\"middle\" x=\"373.5\" y=\"-228.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Phage &#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"373.5\" y=\"-213.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Related</text>\n",
       "</g>\n",
       "<!-- data2_PhoTax -->\n",
       "<g id=\"node24\" class=\"node\">\n",
       "<title>data2_PhoTax</title>\n",
       "<text text-anchor=\"middle\" x=\"535.5\" y=\"-221.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">PhoTax</text>\n",
       "</g>\n",
       "<!-- data2_Purine Metabolism -->\n",
       "<g id=\"node25\" class=\"node\">\n",
       "<title>data2_Purine Metabolism</title>\n",
       "<text text-anchor=\"middle\" x=\"674.5\" y=\"-228.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Purine M&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"674.5\" y=\"-213.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">etabolism</text>\n",
       "</g>\n",
       "<!-- data2_RpaA/RpaB -->\n",
       "<g id=\"node26\" class=\"node\">\n",
       "<title>data2_RpaA/RpaB</title>\n",
       "<text text-anchor=\"middle\" x=\"788.5\" y=\"-221.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">RpaA/RpaB</text>\n",
       "</g>\n",
       "<!-- data2_RpaB -->\n",
       "<g id=\"node27\" class=\"node\">\n",
       "<title>data2_RpaB</title>\n",
       "<text text-anchor=\"middle\" x=\"67.5\" y=\"-118.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">RpaB</text>\n",
       "</g>\n",
       "<!-- data2_SG_1 -->\n",
       "<g id=\"node28\" class=\"node\">\n",
       "<title>data2_SG_1</title>\n",
       "<text text-anchor=\"middle\" x=\"213.5\" y=\"-118.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">SG_1</text>\n",
       "</g>\n",
       "<!-- data2_SG_3 -->\n",
       "<g id=\"node29\" class=\"node\">\n",
       "<title>data2_SG_3</title>\n",
       "<text text-anchor=\"middle\" x=\"373.5\" y=\"-118.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">SG_3</text>\n",
       "</g>\n",
       "<!-- data2_SG_4 -->\n",
       "<g id=\"node30\" class=\"node\">\n",
       "<title>data2_SG_4</title>\n",
       "<text text-anchor=\"middle\" x=\"535.5\" y=\"-118.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">SG_4</text>\n",
       "</g>\n",
       "<!-- data2_SG_8 -->\n",
       "<g id=\"node31\" class=\"node\">\n",
       "<title>data2_SG_8</title>\n",
       "<text text-anchor=\"middle\" x=\"674.5\" y=\"-118.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">SG_8</text>\n",
       "</g>\n",
       "<!-- data2_STR -->\n",
       "<g id=\"node32\" class=\"node\">\n",
       "<title>data2_STR</title>\n",
       "<text text-anchor=\"middle\" x=\"788.5\" y=\"-118.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">STR</text>\n",
       "</g>\n",
       "<!-- data2_SphR -->\n",
       "<g id=\"node33\" class=\"node\">\n",
       "<title>data2_SphR</title>\n",
       "<text text-anchor=\"middle\" x=\"107.5\" y=\"-533.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">SphR</text>\n",
       "</g>\n",
       "<!-- data2_State Transition -->\n",
       "<g id=\"node34\" class=\"node\">\n",
       "<title>data2_State Transition</title>\n",
       "<text text-anchor=\"middle\" x=\"67.5\" y=\"-22.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">State Tr&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"67.5\" y=\"-7.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ansition</text>\n",
       "</g>\n",
       "<!-- data2_Translation -->\n",
       "<g id=\"node35\" class=\"node\">\n",
       "<title>data2_Translation</title>\n",
       "<text text-anchor=\"middle\" x=\"213.5\" y=\"-22.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Trans&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"213.5\" y=\"-7.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">lation</text>\n",
       "</g>\n",
       "<!-- data2_U_11 -->\n",
       "<g id=\"node36\" class=\"node\">\n",
       "<title>data2_U_11</title>\n",
       "<text text-anchor=\"middle\" x=\"332.5\" y=\"-15.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">U_11</text>\n",
       "</g>\n",
       "<!-- data2_U_12 -->\n",
       "<g id=\"node37\" class=\"node\">\n",
       "<title>data2_U_12</title>\n",
       "<text text-anchor=\"middle\" x=\"400.5\" y=\"-15.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">U_12</text>\n",
       "</g>\n",
       "<!-- data2_U_5 -->\n",
       "<g id=\"node38\" class=\"node\">\n",
       "<title>data2_U_5</title>\n",
       "<text text-anchor=\"middle\" x=\"706.5\" y=\"-533.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">U_5</text>\n",
       "</g>\n",
       "<!-- data2_U_7 -->\n",
       "<g id=\"node39\" class=\"node\">\n",
       "<title>data2_U_7</title>\n",
       "<text text-anchor=\"middle\" x=\"535.5\" y=\"-15.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">U_7</text>\n",
       "</g>\n",
       "<!-- data2_U_9 -->\n",
       "<g id=\"node40\" class=\"node\">\n",
       "<title>data2_U_9</title>\n",
       "<text text-anchor=\"middle\" x=\"674.5\" y=\"-15.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">U_9</text>\n",
       "</g>\n",
       "<!-- data2_rpaA/rpopD6  -->\n",
       "<g id=\"node41\" class=\"node\">\n",
       "<title>data2_rpaA/rpopD6 </title>\n",
       "<text text-anchor=\"middle\" x=\"788.5\" y=\"-22.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">rpaA/r&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"788.5\" y=\"-7.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">popD6 </text>\n",
       "</g>\n",
       "<!-- data1_0 -->\n",
       "<g id=\"node42\" class=\"node\">\n",
       "<title>data1_0</title>\n",
       "<text text-anchor=\"middle\" x=\"373.5\" y=\"-591.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">0</text>\n",
       "</g>\n",
       "<!-- data1_0&#45;&gt;data2_CMM -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>data1_0&#45;&gt;data2_CMM</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.82\" d=\"M373.5,-576.89C373.5,-571.57 373.5,-565.63 373.5,-560.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.82\" points=\"375.25,-560 373.5,-555 371.75,-560 375.25,-560\"/>\n",
       "</g>\n",
       "<!-- data1_12 -->\n",
       "<g id=\"node43\" class=\"node\">\n",
       "<title>data1_12</title>\n",
       "<text text-anchor=\"middle\" x=\"788.5\" y=\"-176.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">12</text>\n",
       "</g>\n",
       "<!-- data1_12&#45;&gt;data2_STR -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>data1_12&#45;&gt;data2_STR</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.4\" d=\"M788.5,-161.89C788.5,-156.57 788.5,-150.63 788.5,-145.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.4\" points=\"790.25,-145 788.5,-140 786.75,-145 790.25,-145\"/>\n",
       "</g>\n",
       "<!-- data1_14 -->\n",
       "<g id=\"node44\" class=\"node\">\n",
       "<title>data1_14</title>\n",
       "<text text-anchor=\"middle\" x=\"788.5\" y=\"-279.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">14</text>\n",
       "</g>\n",
       "<!-- data1_14&#45;&gt;data2_RpaA/RpaB -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>data1_14&#45;&gt;data2_RpaA/RpaB</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.17\" d=\"M788.5,-264.89C788.5,-259.57 788.5,-253.63 788.5,-248.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.17\" points=\"790.32,-248 788.5,-243 786.68,-248 790.32,-248\"/>\n",
       "</g>\n",
       "<!-- data1_16 -->\n",
       "<g id=\"node45\" class=\"node\">\n",
       "<title>data1_16</title>\n",
       "<text text-anchor=\"middle\" x=\"213.5\" y=\"-488.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">16</text>\n",
       "</g>\n",
       "<!-- data1_16&#45;&gt;data2_Cytc Oxidases  -->\n",
       "<g id=\"edge13\" class=\"edge\">\n",
       "<title>data1_16&#45;&gt;data2_Cytc Oxidases </title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.13\" d=\"M213.5,-473.9C213.5,-468.68 213.5,-462.84 213.5,-457.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.13\" points=\"215.31,-457.28 213.5,-452.28 211.69,-457.28 215.31,-457.28\"/>\n",
       "</g>\n",
       "<!-- data1_18 -->\n",
       "<g id=\"node46\" class=\"node\">\n",
       "<title>data1_18</title>\n",
       "<text text-anchor=\"middle\" x=\"535.5\" y=\"-487.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">18</text>\n",
       "</g>\n",
       "<!-- data1_18&#45;&gt;data2_Fur&#45;Like -->\n",
       "<g id=\"edge14\" class=\"edge\">\n",
       "<title>data1_18&#45;&gt;data2_Fur&#45;Like</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.67\" d=\"M535.5,-472.89C535.5,-467.57 535.5,-461.63 535.5,-456.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.67\" points=\"537.25,-456 535.5,-451 533.75,-456 537.25,-456\"/>\n",
       "</g>\n",
       "<!-- data1_19 -->\n",
       "<g id=\"node47\" class=\"node\">\n",
       "<title>data1_19</title>\n",
       "<text text-anchor=\"middle\" x=\"67.5\" y=\"-280.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">19</text>\n",
       "</g>\n",
       "<!-- data1_19&#45;&gt;data2_OxPh NADPH ** -->\n",
       "<g id=\"edge15\" class=\"edge\">\n",
       "<title>data1_19&#45;&gt;data2_OxPh NADPH **</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.37\" d=\"M67.5,-265.9C67.5,-260.68 67.5,-254.84 67.5,-249.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.37\" points=\"69.41,-249.28 67.5,-244.28 65.59,-249.28 69.41,-249.28\"/>\n",
       "</g>\n",
       "<!-- data1_2 -->\n",
       "<g id=\"node48\" class=\"node\">\n",
       "<title>data1_2</title>\n",
       "<text text-anchor=\"middle\" x=\"213.5\" y=\"-74.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">2</text>\n",
       "</g>\n",
       "<!-- data1_2&#45;&gt;data2_Translation -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>data1_2&#45;&gt;data2_Translation</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.17\" d=\"M213.5,-59.9C213.5,-54.68 213.5,-48.84 213.5,-43.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.17\" points=\"215.32,-43.28 213.5,-38.28 211.68,-43.28 215.32,-43.28\"/>\n",
       "</g>\n",
       "<!-- data1_20 -->\n",
       "<g id=\"node49\" class=\"node\">\n",
       "<title>data1_20</title>\n",
       "<text text-anchor=\"middle\" x=\"373.5\" y=\"-73.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">20</text>\n",
       "</g>\n",
       "<!-- data1_20&#45;&gt;data2_U_11 -->\n",
       "<g id=\"edge16\" class=\"edge\">\n",
       "<title>data1_20&#45;&gt;data2_U_11</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"1.87\" d=\"M360.98,-58.89C356.92,-53.36 352.38,-47.15 348.13,-41.34\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"1.87\" points=\"349.31,-40 344.95,-37 346.49,-42.07 349.31,-40\"/>\n",
       "</g>\n",
       "<!-- data1_20&#45;&gt;data2_U_12 -->\n",
       "<g id=\"edge17\" class=\"edge\">\n",
       "<title>data1_20&#45;&gt;data2_U_12</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.15\" d=\"M381.75,-58.89C384.36,-53.47 387.29,-47.39 390.04,-41.69\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.15\" points=\"391.71,-42.27 392.3,-37 388.56,-40.75 391.71,-42.27\"/>\n",
       "</g>\n",
       "<!-- data1_22 -->\n",
       "<g id=\"node50\" class=\"node\">\n",
       "<title>data1_22</title>\n",
       "<text text-anchor=\"middle\" x=\"39.5\" y=\"-591.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">22</text>\n",
       "</g>\n",
       "<!-- data1_22&#45;&gt;data2_Biofilm&#45;1 -->\n",
       "<g id=\"edge19\" class=\"edge\">\n",
       "<title>data1_22&#45;&gt;data2_Biofilm&#45;1</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2.45\" d=\"M39.5,-576.89C39.5,-571.57 39.5,-565.63 39.5,-560.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2.45\" points=\"41.25,-560 39.5,-555 37.75,-560 41.25,-560\"/>\n",
       "</g>\n",
       "<!-- data1_23 -->\n",
       "<g id=\"node51\" class=\"node\">\n",
       "<title>data1_23</title>\n",
       "<text text-anchor=\"middle\" x=\"373.5\" y=\"-280.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">23</text>\n",
       "</g>\n",
       "<!-- data1_23&#45;&gt;data2_Phage Related -->\n",
       "<g id=\"edge20\" class=\"edge\">\n",
       "<title>data1_23&#45;&gt;data2_Phage Related</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.54\" d=\"M373.5,-265.9C373.5,-260.68 373.5,-254.84 373.5,-249.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.54\" points=\"375.49,-249.28 373.5,-244.28 371.51,-249.28 375.49,-249.28\"/>\n",
       "</g>\n",
       "<!-- data1_24 -->\n",
       "<g id=\"node52\" class=\"node\">\n",
       "<title>data1_24</title>\n",
       "<text text-anchor=\"middle\" x=\"676.5\" y=\"-592.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">24</text>\n",
       "</g>\n",
       "<!-- data1_24&#45;&gt;data2_Cell Division -->\n",
       "<g id=\"edge21\" class=\"edge\">\n",
       "<title>data1_24&#45;&gt;data2_Cell Division</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"1.82\" d=\"M667.8,-577.9C665.09,-572.57 662.05,-566.6 659.17,-560.94\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"1.82\" points=\"660.63,-559.94 656.8,-556.28 657.51,-561.53 660.63,-559.94\"/>\n",
       "</g>\n",
       "<!-- data1_24&#45;&gt;data2_U_5 -->\n",
       "<g id=\"edge22\" class=\"edge\">\n",
       "<title>data1_24&#45;&gt;data2_U_5</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2.67\" d=\"M685.5,-577.9C688.53,-572.15 691.94,-565.66 695.13,-559.61\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2.67\" points=\"696.73,-560.33 697.51,-555.09 693.63,-558.7 696.73,-560.33\"/>\n",
       "</g>\n",
       "<!-- data1_25 -->\n",
       "<g id=\"node53\" class=\"node\">\n",
       "<title>data1_25</title>\n",
       "<text text-anchor=\"middle\" x=\"67.5\" y=\"-487.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">25</text>\n",
       "</g>\n",
       "<!-- data1_25&#45;&gt;data2_CysR -->\n",
       "<g id=\"edge23\" class=\"edge\">\n",
       "<title>data1_25&#45;&gt;data2_CysR</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.32\" d=\"M67.5,-472.89C67.5,-467.57 67.5,-461.63 67.5,-456.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.32\" points=\"69.39,-456 67.5,-451 65.61,-456 69.39,-456\"/>\n",
       "</g>\n",
       "<!-- data1_26 -->\n",
       "<g id=\"node54\" class=\"node\">\n",
       "<title>data1_26</title>\n",
       "<text text-anchor=\"middle\" x=\"674.5\" y=\"-487.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">26</text>\n",
       "</g>\n",
       "<!-- data1_26&#45;&gt;data2_HSR -->\n",
       "<g id=\"edge24\" class=\"edge\">\n",
       "<title>data1_26&#45;&gt;data2_HSR</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.28\" d=\"M674.5,-472.89C674.5,-467.57 674.5,-461.63 674.5,-456.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.28\" points=\"676.37,-456 674.5,-451 672.63,-456 676.37,-456\"/>\n",
       "</g>\n",
       "<!-- data1_27 -->\n",
       "<g id=\"node55\" class=\"node\">\n",
       "<title>data1_27</title>\n",
       "<text text-anchor=\"middle\" x=\"788.5\" y=\"-487.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">27</text>\n",
       "</g>\n",
       "<!-- data1_27&#45;&gt;data2_IdiB -->\n",
       "<g id=\"edge25\" class=\"edge\">\n",
       "<title>data1_27&#45;&gt;data2_IdiB</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.42\" d=\"M788.5,-472.89C788.5,-467.57 788.5,-461.63 788.5,-456.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.42\" points=\"790.25,-456 788.5,-451 786.75,-456 790.25,-456\"/>\n",
       "</g>\n",
       "<!-- data1_28 -->\n",
       "<g id=\"node56\" class=\"node\">\n",
       "<title>data1_28</title>\n",
       "<text text-anchor=\"middle\" x=\"100.5\" y=\"-591.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">28</text>\n",
       "</g>\n",
       "<!-- data1_28&#45;&gt;data2_Biofilm&#45;1 -->\n",
       "<g id=\"edge27\" class=\"edge\">\n",
       "<title>data1_28&#45;&gt;data2_Biofilm&#45;1</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"1.78\" d=\"M81.87,-576.89C75.6,-571.14 68.54,-564.66 62.01,-558.66\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"1.78\" points=\"62.89,-557.09 58.02,-555 60.52,-559.67 62.89,-557.09\"/>\n",
       "</g>\n",
       "<!-- data1_28&#45;&gt;data2_SphR -->\n",
       "<g id=\"edge26\" class=\"edge\">\n",
       "<title>data1_28&#45;&gt;data2_SphR</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3\" d=\"M102.64,-576.89C103.3,-571.57 104.05,-565.63 104.75,-560.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3\" points=\"106.49,-560.18 105.37,-555 103.02,-559.75 106.49,-560.18\"/>\n",
       "</g>\n",
       "<!-- data1_31 -->\n",
       "<g id=\"node57\" class=\"node\">\n",
       "<title>data1_31</title>\n",
       "<text text-anchor=\"middle\" x=\"67.5\" y=\"-384.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">31</text>\n",
       "</g>\n",
       "<!-- data1_31&#45;&gt;data2_Membrane Biogenesis -->\n",
       "<g id=\"edge30\" class=\"edge\">\n",
       "<title>data1_31&#45;&gt;data2_Membrane Biogenesis</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.81\" d=\"M67.5,-369.9C67.5,-364.68 67.5,-358.84 67.5,-353.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.81\" points=\"69.25,-353.28 67.5,-348.28 65.75,-353.28 69.25,-353.28\"/>\n",
       "</g>\n",
       "<!-- data1_32 -->\n",
       "<g id=\"node58\" class=\"node\">\n",
       "<title>data1_32</title>\n",
       "<text text-anchor=\"middle\" x=\"213.5\" y=\"-591.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">32</text>\n",
       "</g>\n",
       "<!-- data1_32&#45;&gt;data2_Biofilm&#45;2 -->\n",
       "<g id=\"edge31\" class=\"edge\">\n",
       "<title>data1_32&#45;&gt;data2_Biofilm&#45;2</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2.9\" d=\"M213.5,-576.89C213.5,-571.57 213.5,-565.63 213.5,-560.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2.9\" points=\"215.25,-560 213.5,-555 211.75,-560 215.25,-560\"/>\n",
       "</g>\n",
       "<!-- data1_33 -->\n",
       "<g id=\"node59\" class=\"node\">\n",
       "<title>data1_33</title>\n",
       "<text text-anchor=\"middle\" x=\"373.5\" y=\"-488.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">33</text>\n",
       "</g>\n",
       "<!-- data1_33&#45;&gt;data2_Dark&#45;Induced Transition -->\n",
       "<g id=\"edge32\" class=\"edge\">\n",
       "<title>data1_33&#45;&gt;data2_Dark&#45;Induced Transition</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.7\" d=\"M373.5,-473.9C373.5,-468.68 373.5,-462.84 373.5,-457.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.7\" points=\"375.56,-457.28 373.5,-452.28 371.44,-457.28 375.56,-457.28\"/>\n",
       "</g>\n",
       "<!-- data1_34 -->\n",
       "<g id=\"node60\" class=\"node\">\n",
       "<title>data1_34</title>\n",
       "<text text-anchor=\"middle\" x=\"788.5\" y=\"-591.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">34</text>\n",
       "</g>\n",
       "<!-- data1_34&#45;&gt;data2_Comp&#45;2 -->\n",
       "<g id=\"edge33\" class=\"edge\">\n",
       "<title>data1_34&#45;&gt;data2_Comp&#45;2</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.05\" d=\"M788.5,-576.89C788.5,-571.57 788.5,-565.63 788.5,-560.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.05\" points=\"790.27,-560 788.5,-555 786.73,-560 790.27,-560\"/>\n",
       "</g>\n",
       "<!-- data1_36 -->\n",
       "<g id=\"node61\" class=\"node\">\n",
       "<title>data1_36</title>\n",
       "<text text-anchor=\"middle\" x=\"508.5\" y=\"-279.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">36</text>\n",
       "</g>\n",
       "<!-- data1_36&#45;&gt;data2_PhoTax -->\n",
       "<g id=\"edge34\" class=\"edge\">\n",
       "<title>data1_36&#45;&gt;data2_PhoTax</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.06\" d=\"M516.75,-264.89C519.36,-259.47 522.29,-253.39 525.04,-247.69\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.06\" points=\"526.71,-248.27 527.3,-243 523.56,-246.75 526.71,-248.27\"/>\n",
       "</g>\n",
       "<!-- data1_37 -->\n",
       "<g id=\"node62\" class=\"node\">\n",
       "<title>data1_37</title>\n",
       "<text text-anchor=\"middle\" x=\"213.5\" y=\"-279.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">37</text>\n",
       "</g>\n",
       "<!-- data1_37&#45;&gt;data2_OxiTol -->\n",
       "<g id=\"edge35\" class=\"edge\">\n",
       "<title>data1_37&#45;&gt;data2_OxiTol</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.31\" d=\"M213.5,-264.89C213.5,-259.57 213.5,-253.63 213.5,-248.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.31\" points=\"215.25,-248 213.5,-243 211.75,-248 215.25,-248\"/>\n",
       "</g>\n",
       "<!-- data1_38 -->\n",
       "<g id=\"node63\" class=\"node\">\n",
       "<title>data1_38</title>\n",
       "<text text-anchor=\"middle\" x=\"189.5\" y=\"-384.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">38</text>\n",
       "</g>\n",
       "<!-- data1_38&#45;&gt;data2_NblS&#45;RpaB&#45;SrrA&#45;2  -->\n",
       "<g id=\"edge37\" class=\"edge\">\n",
       "<title>data1_38&#45;&gt;data2_NblS&#45;RpaB&#45;SrrA&#45;2 </title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"1.91\" d=\"M188.3,-369.9C187.93,-364.68 187.52,-358.84 187.13,-353.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"1.91\" points=\"188.88,-353.15 186.78,-348.28 185.39,-353.39 188.88,-353.15\"/>\n",
       "</g>\n",
       "<!-- data1_38&#45;&gt;data2_PS -->\n",
       "<g id=\"edge36\" class=\"edge\">\n",
       "<title>data1_38&#45;&gt;data2_PS</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.59\" d=\"M209.6,-369.9C216.76,-363.81 224.89,-356.89 232.35,-350.55\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.59\" points=\"233.74,-351.66 236.41,-347.09 231.47,-349 233.74,-351.66\"/>\n",
       "</g>\n",
       "<!-- data1_42 -->\n",
       "<g id=\"node64\" class=\"node\">\n",
       "<title>data1_42</title>\n",
       "<text text-anchor=\"middle\" x=\"674.5\" y=\"-73.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">42</text>\n",
       "</g>\n",
       "<!-- data1_42&#45;&gt;data2_U_9 -->\n",
       "<g id=\"edge40\" class=\"edge\">\n",
       "<title>data1_42&#45;&gt;data2_U_9</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.47\" d=\"M674.5,-58.89C674.5,-53.57 674.5,-47.63 674.5,-42.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.47\" points=\"676.46,-42 674.5,-37 672.54,-42 676.46,-42\"/>\n",
       "</g>\n",
       "<!-- data1_44 -->\n",
       "<g id=\"node65\" class=\"node\">\n",
       "<title>data1_44</title>\n",
       "<text text-anchor=\"middle\" x=\"67.5\" y=\"-74.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">44</text>\n",
       "</g>\n",
       "<!-- data1_44&#45;&gt;data2_State Transition -->\n",
       "<g id=\"edge42\" class=\"edge\">\n",
       "<title>data1_44&#45;&gt;data2_State Transition</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.12\" d=\"M67.5,-59.9C67.5,-54.68 67.5,-48.84 67.5,-43.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.12\" points=\"69.3,-43.28 67.5,-38.28 65.7,-43.28 69.3,-43.28\"/>\n",
       "</g>\n",
       "<!-- data1_48 -->\n",
       "<g id=\"node66\" class=\"node\">\n",
       "<title>data1_48</title>\n",
       "<text text-anchor=\"middle\" x=\"674.5\" y=\"-383.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">48</text>\n",
       "</g>\n",
       "<!-- data1_48&#45;&gt;data2_NtcA&#45;3 -->\n",
       "<g id=\"edge43\" class=\"edge\">\n",
       "<title>data1_48&#45;&gt;data2_NtcA&#45;3</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.21\" d=\"M674.5,-368.89C674.5,-363.57 674.5,-357.63 674.5,-352.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.21\" points=\"676.25,-352 674.5,-347 672.75,-352 676.25,-352\"/>\n",
       "</g>\n",
       "<!-- data1_5 -->\n",
       "<g id=\"node67\" class=\"node\">\n",
       "<title>data1_5</title>\n",
       "<text text-anchor=\"middle\" x=\"67.5\" y=\"-176.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">5</text>\n",
       "</g>\n",
       "<!-- data1_5&#45;&gt;data2_RpaB -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>data1_5&#45;&gt;data2_RpaB</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.95\" d=\"M67.5,-161.89C67.5,-156.57 67.5,-150.63 67.5,-145.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.95\" points=\"69.25,-145 67.5,-140 65.75,-145 69.25,-145\"/>\n",
       "</g>\n",
       "<!-- data1_51 -->\n",
       "<g id=\"node68\" class=\"node\">\n",
       "<title>data1_51</title>\n",
       "<text text-anchor=\"middle\" x=\"428.5\" y=\"-73.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">51</text>\n",
       "</g>\n",
       "<!-- data1_51&#45;&gt;data2_U_12 -->\n",
       "<g id=\"edge44\" class=\"edge\">\n",
       "<title>data1_51&#45;&gt;data2_U_12</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"1.85\" d=\"M419.95,-58.89C417.23,-53.47 414.2,-47.39 411.34,-41.69\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"1.85\" points=\"412.8,-40.69 409,-37 409.67,-42.26 412.8,-40.69\"/>\n",
       "</g>\n",
       "<!-- data1_52 -->\n",
       "<g id=\"node69\" class=\"node\">\n",
       "<title>data1_52</title>\n",
       "<text text-anchor=\"middle\" x=\"674.5\" y=\"-176.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">52</text>\n",
       "</g>\n",
       "<!-- data1_52&#45;&gt;data2_SG_8 -->\n",
       "<g id=\"edge45\" class=\"edge\">\n",
       "<title>data1_52&#45;&gt;data2_SG_8</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2.77\" d=\"M674.5,-161.89C674.5,-156.57 674.5,-150.63 674.5,-145.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2.77\" points=\"676.25,-145 674.5,-140 672.75,-145 676.25,-145\"/>\n",
       "</g>\n",
       "<!-- data1_53 -->\n",
       "<g id=\"node70\" class=\"node\">\n",
       "<title>data1_53</title>\n",
       "<text text-anchor=\"middle\" x=\"318.5\" y=\"-73.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">53</text>\n",
       "</g>\n",
       "<!-- data1_53&#45;&gt;data2_U_11 -->\n",
       "<g id=\"edge46\" class=\"edge\">\n",
       "<title>data1_53&#45;&gt;data2_U_11</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.85\" d=\"M322.78,-58.89C324.11,-53.57 325.59,-47.63 326.99,-42.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.85\" points=\"328.73,-42.28 328.25,-37 325.34,-41.43 328.73,-42.28\"/>\n",
       "</g>\n",
       "<!-- data1_55 -->\n",
       "<g id=\"node71\" class=\"node\">\n",
       "<title>data1_55</title>\n",
       "<text text-anchor=\"middle\" x=\"788.5\" y=\"-383.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">55</text>\n",
       "</g>\n",
       "<!-- data1_55&#45;&gt;data2_NtcA&#45;4 -->\n",
       "<g id=\"edge47\" class=\"edge\">\n",
       "<title>data1_55&#45;&gt;data2_NtcA&#45;4</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2.85\" d=\"M788.5,-368.89C788.5,-363.57 788.5,-357.63 788.5,-352.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2.85\" points=\"790.25,-352 788.5,-347 786.75,-352 790.25,-352\"/>\n",
       "</g>\n",
       "<!-- data1_6 -->\n",
       "<g id=\"node72\" class=\"node\">\n",
       "<title>data1_6</title>\n",
       "<text text-anchor=\"middle\" x=\"252.5\" y=\"-384.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">6</text>\n",
       "</g>\n",
       "<!-- data1_6&#45;&gt;data2_NblS&#45;RpaB&#45;SrrA&#45;2  -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>data1_6&#45;&gt;data2_NblS&#45;RpaB&#45;SrrA&#45;2 </title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.36\" d=\"M232.4,-369.9C225.63,-364.14 218,-357.65 210.88,-351.59\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.36\" points=\"211.93,-350.19 206.99,-348.28 209.66,-352.85 211.93,-350.19\"/>\n",
       "</g>\n",
       "<!-- data1_6&#45;&gt;data2_PS -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>data1_6&#45;&gt;data2_PS</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2.13\" d=\"M253.7,-369.9C254.09,-364.37 254.52,-358.17 254.93,-352.32\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2.13\" points=\"256.7,-352.2 255.3,-347.09 253.2,-351.96 256.7,-352.2\"/>\n",
       "</g>\n",
       "<!-- data1_7 -->\n",
       "<g id=\"node73\" class=\"node\">\n",
       "<title>data1_7</title>\n",
       "<text text-anchor=\"middle\" x=\"674.5\" y=\"-280.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">7</text>\n",
       "</g>\n",
       "<!-- data1_7&#45;&gt;data2_Purine Metabolism -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>data1_7&#45;&gt;data2_Purine Metabolism</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.93\" d=\"M674.5,-265.9C674.5,-260.68 674.5,-254.84 674.5,-249.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.93\" points=\"676.25,-249.28 674.5,-244.28 672.75,-249.28 676.25,-249.28\"/>\n",
       "</g>\n",
       "<!-- data1_8 -->\n",
       "<g id=\"node74\" class=\"node\">\n",
       "<title>data1_8</title>\n",
       "<text text-anchor=\"middle\" x=\"535.5\" y=\"-73.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">8</text>\n",
       "</g>\n",
       "<!-- data1_8&#45;&gt;data2_U_7 -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>data1_8&#45;&gt;data2_U_7</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.23\" d=\"M535.5,-58.89C535.5,-53.57 535.5,-47.63 535.5,-42.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.23\" points=\"537.35,-42 535.5,-37 533.65,-42 537.35,-42\"/>\n",
       "</g>\n",
       "<!-- data1_SG_10 -->\n",
       "<g id=\"node75\" class=\"node\">\n",
       "<title>data1_SG_10</title>\n",
       "<text text-anchor=\"middle\" x=\"788.5\" y=\"-74.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">SG_10</text>\n",
       "</g>\n",
       "<!-- data1_SG_10&#45;&gt;data2_rpaA/rpopD6  -->\n",
       "<g id=\"edge41\" class=\"edge\">\n",
       "<title>data1_SG_10&#45;&gt;data2_rpaA/rpopD6 </title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.17\" d=\"M788.5,-59.9C788.5,-54.68 788.5,-48.84 788.5,-43.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.17\" points=\"790.32,-43.28 788.5,-38.28 786.68,-43.28 790.32,-43.28\"/>\n",
       "</g>\n",
       "<!-- data1_SG_3 -->\n",
       "<g id=\"node76\" class=\"node\">\n",
       "<title>data1_SG_3</title>\n",
       "<text text-anchor=\"middle\" x=\"373.5\" y=\"-176.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">SG_3</text>\n",
       "</g>\n",
       "<!-- data1_SG_3&#45;&gt;data2_SG_3 -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>data1_SG_3&#45;&gt;data2_SG_3</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.08\" d=\"M373.5,-161.89C373.5,-156.57 373.5,-150.63 373.5,-145.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.08\" points=\"375.29,-145 373.5,-140 371.72,-145 375.29,-145\"/>\n",
       "</g>\n",
       "<!-- data1_SG_5 -->\n",
       "<g id=\"node77\" class=\"node\">\n",
       "<title>data1_SG_5</title>\n",
       "<text text-anchor=\"middle\" x=\"563.5\" y=\"-279.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">SG_5</text>\n",
       "</g>\n",
       "<!-- data1_SG_5&#45;&gt;data2_PhoTax -->\n",
       "<g id=\"edge28\" class=\"edge\">\n",
       "<title>data1_SG_5&#45;&gt;data2_PhoTax</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"1.75\" d=\"M554.95,-264.89C552.23,-259.47 549.2,-253.39 546.34,-247.69\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"1.75\" points=\"547.8,-246.69 544,-243 544.67,-248.26 547.8,-246.69\"/>\n",
       "</g>\n",
       "<!-- data1_SG_6 -->\n",
       "<g id=\"node78\" class=\"node\">\n",
       "<title>data1_SG_6</title>\n",
       "<text text-anchor=\"middle\" x=\"213.5\" y=\"-176.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">SG_6</text>\n",
       "</g>\n",
       "<!-- data1_SG_6&#45;&gt;data2_SG_1 -->\n",
       "<g id=\"edge29\" class=\"edge\">\n",
       "<title>data1_SG_6&#45;&gt;data2_SG_1</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2.53\" d=\"M213.5,-161.89C213.5,-156.57 213.5,-150.63 213.5,-145.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2.53\" points=\"215.25,-145 213.5,-140 211.75,-145 215.25,-145\"/>\n",
       "</g>\n",
       "<!-- data1_SG_8 -->\n",
       "<g id=\"node79\" class=\"node\">\n",
       "<title>data1_SG_8</title>\n",
       "<text text-anchor=\"middle\" x=\"535.5\" y=\"-176.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">SG_8</text>\n",
       "</g>\n",
       "<!-- data1_SG_8&#45;&gt;data2_SG_4 -->\n",
       "<g id=\"edge38\" class=\"edge\">\n",
       "<title>data1_SG_8&#45;&gt;data2_SG_4</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2.83\" d=\"M535.5,-161.89C535.5,-156.57 535.5,-150.63 535.5,-145.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2.83\" points=\"537.25,-145 535.5,-140 533.75,-145 537.25,-145\"/>\n",
       "</g>\n",
       "<!-- data1_ccmR&#45;1 -->\n",
       "<g id=\"node80\" class=\"node\">\n",
       "<title>data1_ccmR&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"541.5\" y=\"-592.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ccmR&#45;1</text>\n",
       "</g>\n",
       "<!-- data1_ccmR&#45;1&#45;&gt;data2_Carbon Transport -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>data1_ccmR&#45;1&#45;&gt;data2_Carbon Transport</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2.85\" d=\"M530.7,-577.9C527.27,-572.46 523.41,-566.37 519.78,-560.6\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2.85\" points=\"521.19,-559.57 517.05,-556.28 518.24,-561.44 521.19,-559.57\"/>\n",
       "</g>\n",
       "<!-- data1_ccmR&#45;1&#45;&gt;data2_CmpR -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>data1_ccmR&#45;1&#45;&gt;data2_CmpR</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.47\" d=\"M552.6,-577.9C556.33,-572.15 560.55,-565.66 564.47,-559.61\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.47\" points=\"566.15,-560.24 567.41,-555.09 563.22,-558.33 566.15,-560.24\"/>\n",
       "</g>\n",
       "<!-- data1_ntcA&#45;1 -->\n",
       "<g id=\"node81\" class=\"node\">\n",
       "<title>data1_ntcA&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"373.5\" y=\"-383.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ntcA&#45;1</text>\n",
       "</g>\n",
       "<!-- data1_ntcA&#45;1&#45;&gt;data2_NtcA&#45;1 -->\n",
       "<g id=\"edge18\" class=\"edge\">\n",
       "<title>data1_ntcA&#45;1&#45;&gt;data2_NtcA&#45;1</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.21\" d=\"M373.5,-368.89C373.5,-363.57 373.5,-357.63 373.5,-352.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.21\" points=\"375.34,-352 373.5,-347 371.66,-352 375.34,-352\"/>\n",
       "</g>\n",
       "<!-- data1_ntcA&#45;2 -->\n",
       "<g id=\"node82\" class=\"node\">\n",
       "<title>data1_ntcA&#45;2</title>\n",
       "<text text-anchor=\"middle\" x=\"535.5\" y=\"-383.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ntcA&#45;2</text>\n",
       "</g>\n",
       "<!-- data1_ntcA&#45;2&#45;&gt;data2_NtcA&#45;2 -->\n",
       "<g id=\"edge39\" class=\"edge\">\n",
       "<title>data1_ntcA&#45;2&#45;&gt;data2_NtcA&#45;2</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.39\" d=\"M535.5,-368.89C535.5,-363.57 535.5,-357.63 535.5,-352.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.39\" points=\"537.42,-352 535.5,-347 533.58,-352 537.42,-352\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.dot.Digraph at 0x7f01176edf40>"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dots,links = compare_ica(ica_data.M,old.M,cutoff=0.3,\n",
    "                         ortho_file = mapping)\n",
    "links"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "hits_dict = dict((x[0], x[1]) for x in dots)\n",
    "hits = [x[0]for x in dots]\n",
    "for i in ica_data.imodulon_names:\n",
    "    if 'SG_' not in str(i) and i in hits and hits.count(i) == 1:\n",
    "        hit_imod = hits_dict[i]\n",
    "        if 'SG' not in hit_imod and 'U_' not in hit_imod:\n",
    "            ica_data.rename_imodulons({i:hit_imod})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save iModulon object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon.util import explained_variance\n",
    "from pymodulon.io import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add iModulon sizes and explained variance\n",
    "for im in ica_data.imodulon_names:\n",
    "    ica_data.imodulon_table.loc[im,'imodulon_size'] = len(ica_data.view_imodulon(im))\n",
    "    ica_data.imodulon_table.loc[im,'explained_variance'] = explained_variance(ica_data,imodulons=im)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This will save your iModulon table, your thresholds, and any other information stored in the ica_data object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'../data/processed_data'"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_to_json(ica_data, path.join(data_dir,'ica_wip.json.gz'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you prefer to view and edit your iModulon table in excel, save it as a CSV and reload the iModulon as before"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data.imodulon_table.to_csv(path.join(data_dir,'imodulon_table_raw.csv'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.1"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "288px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
