{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Setup\" data-toc-modified-id=\"Setup-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Setup</a></span><ul class=\"toc-item\"><li><span><a href=\"#Check-your-sample-table-(i.e.-metadata-file)\" data-toc-modified-id=\"Check-your-sample-table-(i.e.-metadata-file)-1.1\"><span class=\"toc-item-num\">1.1&nbsp;&nbsp;</span>Check your sample table (i.e. metadata file)</a></span></li><li><span><a href=\"#Check-your-TRN\" data-toc-modified-id=\"Check-your-TRN-1.2\"><span class=\"toc-item-num\">1.2&nbsp;&nbsp;</span>Check your TRN</a></span></li><li><span><a href=\"#Load-the-data\" data-toc-modified-id=\"Load-the-data-1.3\"><span class=\"toc-item-num\">1.3&nbsp;&nbsp;</span>Load the data</a></span></li></ul></li><li><span><a href=\"#Regulatory-iModulons\" data-toc-modified-id=\"Regulatory-iModulons-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Regulatory iModulons</a></span></li><li><span><a href=\"#Functional-iModulons\" data-toc-modified-id=\"Functional-iModulons-3\"><span class=\"toc-item-num\">3&nbsp;&nbsp;</span>Functional iModulons</a></span><ul class=\"toc-item\"><li><span><a href=\"#GO-Enrichments\" data-toc-modified-id=\"GO-Enrichments-3.1\"><span class=\"toc-item-num\">3.1&nbsp;&nbsp;</span>GO Enrichments</a></span></li><li><span><a href=\"#KEGG-Enrichments\" data-toc-modified-id=\"KEGG-Enrichments-3.2\"><span class=\"toc-item-num\">3.2&nbsp;&nbsp;</span>KEGG Enrichments</a></span><ul class=\"toc-item\"><li><span><a href=\"#Load-KEGG-mapping\" data-toc-modified-id=\"Load-KEGG-mapping-3.2.1\"><span class=\"toc-item-num\">3.2.1&nbsp;&nbsp;</span>Load KEGG mapping</a></span></li><li><span><a href=\"#Perform-enrichment\" data-toc-modified-id=\"Perform-enrichment-3.2.2\"><span class=\"toc-item-num\">3.2.2&nbsp;&nbsp;</span>Perform enrichment</a></span></li><li><span><a href=\"#Convert-KEGG-IDs-to-human-readable-names\" data-toc-modified-id=\"Convert-KEGG-IDs-to-human-readable-names-3.2.3\"><span class=\"toc-item-num\">3.2.3&nbsp;&nbsp;</span>Convert KEGG IDs to human-readable names</a></span></li></ul></li><li><span><a href=\"#Save-files\" data-toc-modified-id=\"Save-files-3.3\"><span class=\"toc-item-num\">3.3&nbsp;&nbsp;</span>Save files</a></span></li></ul></li><li><span><a href=\"#Check-for-single-gene-iModulons\" data-toc-modified-id=\"Check-for-single-gene-iModulons-4\"><span class=\"toc-item-num\">4&nbsp;&nbsp;</span>Check for single gene iModulons</a></span></li><li><span><a href=\"#Old-characterizations\" data-toc-modified-id=\"Old-characterizations-5\"><span class=\"toc-item-num\">5&nbsp;&nbsp;</span>Old characterizations</a></span></li><li><span><a href=\"#Save-iModulon-object\" data-toc-modified-id=\"Save-iModulon-object-6\"><span class=\"toc-item-num\">6&nbsp;&nbsp;</span>Save iModulon object</a></span></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup\n",
    "This IPython notebook will walk through the steps of characterizing iModulons through the semi-automated tools in PyModulon. You will need:\n",
    "\n",
    "* M and A matrices\n",
    "* Expression data (e.g. `log_tpm_norm.csv`)\n",
    "* Gene table and KEGG/GO annotations (Generated in `gene_annotation.ipynb`)\n",
    "* Sample table, with a column for `project` and `condition`\n",
    "* TRN file\n",
    "\n",
    "Optional:\n",
    "* iModulon table (if you already have some characterized iModulons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "from os import path\n",
    "import sys\n",
    "sys.path.append('/home/annie/Documents/multi-Ecoli/strains/')\n",
    "from pymodulon.core import IcaData\n",
    "from pymodulon.plotting import *\n",
    "import pandas as pd\n",
    "import re\n",
    "from Bio.KEGG import REST\n",
    "from tqdm.notebook import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Enter the location of your data here\n",
    "data_dir = path.join('..','data','all_strict')\n",
    "\n",
    "# GO and KEGG annotations are in the 'external' folder\n",
    "external_data = path.join('..','data','external')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check your sample table (i.e. metadata file)\n",
    "Your metadata file will probably have a lot of columns, most of which you may not care about. Feel free to save a secondary copy of your metadata file with only columns that seem relevant to you. The two most important columns are:\n",
    "1. `project`\n",
    "2. `condition`\n",
    "\n",
    "Make sure that these columns exist in your metadata file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project</th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Experiment</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRX10864008</th>\n",
       "      <td>photomixotrophic</td>\n",
       "      <td>photoautotroph</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX10864009</th>\n",
       "      <td>photomixotrophic</td>\n",
       "      <td>photoautotroph</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX10864010</th>\n",
       "      <td>photomixotrophic</td>\n",
       "      <td>photoautotroph</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX10864011</th>\n",
       "      <td>photomixotrophic</td>\n",
       "      <td>glucose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX10864012</th>\n",
       "      <td>photomixotrophic</td>\n",
       "      <td>glucose</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      project       condition\n",
       "Experiment                                   \n",
       "SRX10864008  photomixotrophic  photoautotroph\n",
       "SRX10864009  photomixotrophic  photoautotroph\n",
       "SRX10864010  photomixotrophic  photoautotroph\n",
       "SRX10864011  photomixotrophic         glucose\n",
       "SRX10864012  photomixotrophic         glucose"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metadata = pd.read_csv(path.join(data_dir,'metadata.tsv'),index_col=0,sep='\\t')\n",
    "df_metadata[['project','condition']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "print(df_metadata.project.notnull().all())\n",
    "print(df_metadata.condition.notnull().all())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check your TRN"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each row of the TRN file represents a regulatory interaction.  \n",
    "**Your TRN file must have the following columns:**\n",
    "1. `regulator` - Name of regulator (`/` or `+` characters will be converted to `;`)\n",
    "1. `gene_id` - Locus tag of gene being regulated\n",
    "\n",
    "The following columns are optional, but are helpful to have:\n",
    "1. `regulator_id` - Locus tag of regulator\n",
    "1. `gene_name` - Name of gene (can automatically update this using `name2num`)\n",
    "1. `direction` - Direction of regulation ('+' for activation, '-' for repression, '?' or NaN for unknown)\n",
    "1. `evidence` - Evidence of regulation (e.g. ChIP-exo, qRT-PCR, SELEX, Motif search)\n",
    "1. `PMID` - Reference for regulation\n",
    "\n",
    "You may add any other columns that could help you. TRNs may be saved as either CSV or TSV files. See below for an example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator_id</th>\n",
       "      <th>regulator</th>\n",
       "      <th>repressor</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>old_locus_tag</th>\n",
       "      <th>gene_product</th>\n",
       "      <th>COG</th>\n",
       "      <th>Direction</th>\n",
       "      <th>Operon_based_assumption</th>\n",
       "      <th>Evidence</th>\n",
       "      <th>PMID</th>\n",
       "      <th>Additional Notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SYNPCC7942_RS00005</td>\n",
       "      <td>SYNPCC7942_RS00005</td>\n",
       "      <td>Synpcc7942_0001</td>\n",
       "      <td>DNA polymerase III subunit beta</td>\n",
       "      <td>Replication and repair</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SYNPCC7942_RS00645</td>\n",
       "      <td>NtcA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ntcA</td>\n",
       "      <td>SYNPCC7942_RS00645</td>\n",
       "      <td>Synpcc7942_0127</td>\n",
       "      <td>global nitrogen regulator NtcA</td>\n",
       "      <td>Transcription</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>pdhB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SYNPCC7942_RS00725</td>\n",
       "      <td>SYNPCC7942_RS00725</td>\n",
       "      <td>Synpcc7942_0143</td>\n",
       "      <td>alpha-ketoacid dehydrogenase subunit beta</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SYNPCC7942_RS00645</td>\n",
       "      <td>NtcA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>glnN</td>\n",
       "      <td>SYNPCC7942_RS00860</td>\n",
       "      <td>Synpcc7942_0169</td>\n",
       "      <td>glutamine synthetase</td>\n",
       "      <td>Function Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ggt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SYNPCC7942_RS00940</td>\n",
       "      <td>SYNPCC7942_RS00940</td>\n",
       "      <td>Synpcc7942_0185</td>\n",
       "      <td>gamma-glutamyltranspeptidase</td>\n",
       "      <td>Amino Acid metabolis and transport</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         regulator_id regulator repressor           gene_name  \\\n",
       "0                 NaN      dnaN       NaN  SYNPCC7942_RS00005   \n",
       "1  SYNPCC7942_RS00645      NtcA       NaN                ntcA   \n",
       "2                 NaN      pdhB       NaN  SYNPCC7942_RS00725   \n",
       "3  SYNPCC7942_RS00645      NtcA       NaN                glnN   \n",
       "4                 NaN       ggt       NaN  SYNPCC7942_RS00940   \n",
       "\n",
       "              gene_id    old_locus_tag  \\\n",
       "0  SYNPCC7942_RS00005  Synpcc7942_0001   \n",
       "1  SYNPCC7942_RS00645  Synpcc7942_0127   \n",
       "2  SYNPCC7942_RS00725  Synpcc7942_0143   \n",
       "3  SYNPCC7942_RS00860  Synpcc7942_0169   \n",
       "4  SYNPCC7942_RS00940  Synpcc7942_0185   \n",
       "\n",
       "                                gene_product  \\\n",
       "0            DNA polymerase III subunit beta   \n",
       "1             global nitrogen regulator NtcA   \n",
       "2  alpha-ketoacid dehydrogenase subunit beta   \n",
       "3                       glutamine synthetase   \n",
       "4               gamma-glutamyltranspeptidase   \n",
       "\n",
       "                                  COG Direction Operon_based_assumption  \\\n",
       "0              Replication and repair       NaN                     NaN   \n",
       "1                       Transcription       NaN                     NaN   \n",
       "2    Energy production and conversion       NaN                     NaN   \n",
       "3                    Function Unknown       NaN                     NaN   \n",
       "4  Amino Acid metabolis and transport       NaN                     NaN   \n",
       "\n",
       "  Evidence PMID Additional Notes  \n",
       "0      NaN  NaN              NaN  \n",
       "1      NaN  NaN              NaN  \n",
       "2      NaN  NaN              NaN  \n",
       "3      NaN  NaN              NaN  \n",
       "4      NaN  NaN              NaN  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_trn = pd.read_csv(path.join(external_data,'TRN.csv'))\n",
    "df_trn.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `regulator` and `gene_id` must be filled in for each row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "print(df_trn.regulator.notnull().all())\n",
    "print(df_trn.gene_id.notnull().all())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load the data\n",
    "You're now ready to load your IcaData object!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:Using Kmeans threshold method. D'agostino optimization will not be performed.\n"
     ]
    }
   ],
   "source": [
    "ica_data = IcaData(M = path.join('../data/all_strict/dimensions/120_optimal/','S.csv'),\n",
    "                   A = path.join('../data/all_strict/dimensions/120_optimal/','A.csv'),\n",
    "                   X = path.join(data_dir,'log_tpm_norm.csv'),\n",
    "                   gene_table = path.join('../data/processed_data/','gene_info.csv'),\n",
    "                   sample_table = path.join(data_dir,'metadata.tsv'),\n",
    "#                    trn = path.join(external_data,'TRN.csv'),\n",
    "                   optimize_cutoff=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you don't have a TRN (or have a very minimal TRN), use `threshold_method = 'kmeans'`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ica_data = IcaData(M = path.join(data_dir,'M.csv'),\n",
    "#                    A = path.join(data_dir,'A.csv'),\n",
    "#                    X = path.join(data_dir,'log_tpm_norm.csv'),\n",
    "#                    gene_table = path.join(data_dir,'gene_info.csv'),\n",
    "#                    sample_table = path.join(data_dir,'metadata.tsv'),\n",
    "#                    trn = path.join(data_dir,'TRN.csv'),\n",
    "#                    threshold_method = 'kmeans')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Regulatory iModulons\n",
    "Use `compute_trn_enrichment` to automatically check for Regulatory iModulons. The more complete your TRN, the more regulatory iModulons you'll find."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "ica_data.compute_trn_enrichment()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can also search for AND/OR combinations of regulators using the `max_regs` argument.\n",
    "\n",
    "Regulator enrichments can be directly saved to the `imodulon_table` using the `save` argument. This saves the enrichment with the lowest q-value to the table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>TrpR</td>\n",
       "      <td>1.167011e-19</td>\n",
       "      <td>1.167011e-19</td>\n",
       "      <td>0.692308</td>\n",
       "      <td>0.818182</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7</td>\n",
       "      <td>FeHm</td>\n",
       "      <td>1.385089e-10</td>\n",
       "      <td>1.385089e-09</td>\n",
       "      <td>0.375000</td>\n",
       "      <td>0.159574</td>\n",
       "      <td>0.223881</td>\n",
       "      <td>30.0</td>\n",
       "      <td>188.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9</td>\n",
       "      <td>NimR</td>\n",
       "      <td>2.691412e-07</td>\n",
       "      <td>2.691412e-07</td>\n",
       "      <td>0.375000</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11</td>\n",
       "      <td>CRP</td>\n",
       "      <td>1.599720e-17</td>\n",
       "      <td>1.599720e-17</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.526316</td>\n",
       "      <td>10.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11</td>\n",
       "      <td>Sxy</td>\n",
       "      <td>1.599720e-17</td>\n",
       "      <td>1.599720e-17</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.526316</td>\n",
       "      <td>10.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>14</td>\n",
       "      <td>FeHm</td>\n",
       "      <td>3.036741e-21</td>\n",
       "      <td>1.214696e-20</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>0.148936</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>28.0</td>\n",
       "      <td>188.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>14</td>\n",
       "      <td>Fur</td>\n",
       "      <td>7.154703e-13</td>\n",
       "      <td>1.430941e-12</td>\n",
       "      <td>0.277778</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.357143</td>\n",
       "      <td>10.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>16</td>\n",
       "      <td>NarP</td>\n",
       "      <td>2.127768e-24</td>\n",
       "      <td>6.383303e-24</td>\n",
       "      <td>0.435897</td>\n",
       "      <td>0.680000</td>\n",
       "      <td>0.531250</td>\n",
       "      <td>17.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>16</td>\n",
       "      <td>FeHm</td>\n",
       "      <td>2.459017e-09</td>\n",
       "      <td>3.688525e-09</td>\n",
       "      <td>0.487179</td>\n",
       "      <td>0.101064</td>\n",
       "      <td>0.167401</td>\n",
       "      <td>19.0</td>\n",
       "      <td>188.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>18</td>\n",
       "      <td>PurR</td>\n",
       "      <td>6.466229e-18</td>\n",
       "      <td>5.819606e-17</td>\n",
       "      <td>0.244898</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>0.380952</td>\n",
       "      <td>12.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>19</td>\n",
       "      <td>MetR</td>\n",
       "      <td>3.339324e-08</td>\n",
       "      <td>2.337527e-07</td>\n",
       "      <td>0.235294</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>23</td>\n",
       "      <td>CRP</td>\n",
       "      <td>4.472261e-31</td>\n",
       "      <td>2.012517e-30</td>\n",
       "      <td>0.306667</td>\n",
       "      <td>0.920000</td>\n",
       "      <td>0.460000</td>\n",
       "      <td>23.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>23</td>\n",
       "      <td>Sxy</td>\n",
       "      <td>4.472261e-31</td>\n",
       "      <td>2.012517e-30</td>\n",
       "      <td>0.306667</td>\n",
       "      <td>0.920000</td>\n",
       "      <td>0.460000</td>\n",
       "      <td>23.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>29</td>\n",
       "      <td>ArgR</td>\n",
       "      <td>2.056349e-16</td>\n",
       "      <td>4.112697e-16</td>\n",
       "      <td>0.538462</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.700000</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>36</td>\n",
       "      <td>CRP</td>\n",
       "      <td>1.210653e-07</td>\n",
       "      <td>3.026632e-07</td>\n",
       "      <td>0.225806</td>\n",
       "      <td>0.280000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>7.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>36</td>\n",
       "      <td>Sxy</td>\n",
       "      <td>1.210653e-07</td>\n",
       "      <td>3.026632e-07</td>\n",
       "      <td>0.225806</td>\n",
       "      <td>0.280000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>7.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    imodulon regulator        pvalue        qvalue  precision    recall  \\\n",
       "0          0      TrpR  1.167011e-19  1.167011e-19   0.692308  0.818182   \n",
       "1          7      FeHm  1.385089e-10  1.385089e-09   0.375000  0.159574   \n",
       "2          9      NimR  2.691412e-07  2.691412e-07   0.375000  0.750000   \n",
       "3         11       CRP  1.599720e-17  1.599720e-17   0.769231  0.400000   \n",
       "4         11       Sxy  1.599720e-17  1.599720e-17   0.769231  0.400000   \n",
       "5         14      FeHm  3.036741e-21  1.214696e-20   0.777778  0.148936   \n",
       "6         14       Fur  7.154703e-13  1.430941e-12   0.277778  0.500000   \n",
       "7         16      NarP  2.127768e-24  6.383303e-24   0.435897  0.680000   \n",
       "8         16      FeHm  2.459017e-09  3.688525e-09   0.487179  0.101064   \n",
       "9         18      PurR  6.466229e-18  5.819606e-17   0.244898  0.857143   \n",
       "10        19      MetR  3.339324e-08  2.337527e-07   0.235294  0.800000   \n",
       "11        23       CRP  4.472261e-31  2.012517e-30   0.306667  0.920000   \n",
       "12        23       Sxy  4.472261e-31  2.012517e-30   0.306667  0.920000   \n",
       "13        29      ArgR  2.056349e-16  4.112697e-16   0.538462  1.000000   \n",
       "14        36       CRP  1.210653e-07  3.026632e-07   0.225806  0.280000   \n",
       "15        36       Sxy  1.210653e-07  3.026632e-07   0.225806  0.280000   \n",
       "\n",
       "     f1score    TP  regulon_size  imodulon_size  n_regs  \n",
       "0   0.750000   9.0          11.0           13.0     1.0  \n",
       "1   0.223881  30.0         188.0           80.0     1.0  \n",
       "2   0.500000   3.0           4.0            8.0     1.0  \n",
       "3   0.526316  10.0          25.0           13.0     1.0  \n",
       "4   0.526316  10.0          25.0           13.0     1.0  \n",
       "5   0.250000  28.0         188.0           36.0     1.0  \n",
       "6   0.357143  10.0          20.0           36.0     1.0  \n",
       "7   0.531250  17.0          25.0           39.0     1.0  \n",
       "8   0.167401  19.0         188.0           39.0     1.0  \n",
       "9   0.380952  12.0          14.0           49.0     1.0  \n",
       "10  0.363636   4.0           5.0           17.0     1.0  \n",
       "11  0.460000  23.0          25.0           75.0     1.0  \n",
       "12  0.460000  23.0          25.0           75.0     1.0  \n",
       "13  0.700000   7.0           7.0           13.0     1.0  \n",
       "14  0.250000   7.0          25.0           31.0     1.0  \n",
       "15  0.250000   7.0          25.0           31.0     1.0  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# First search for regulator enrichments with 2 regulators\n",
    "ica_data.compute_trn_enrichment(max_regs=2,save=True)\n",
    "\n",
    "# Next, search for regulator enrichments with just one regulator. This will supercede the 2 regulator enrichments.\n",
    "ica_data.compute_trn_enrichment(max_regs=1,save=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The list of regulatory iModulons are shown below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45 Total iModulons\n",
      "11 Regulatory iModulons\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TrpR</td>\n",
       "      <td>1.167011e-19</td>\n",
       "      <td>1.167011e-19</td>\n",
       "      <td>0.692308</td>\n",
       "      <td>0.818182</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>FeHm</td>\n",
       "      <td>1.385089e-10</td>\n",
       "      <td>1.385089e-09</td>\n",
       "      <td>0.375000</td>\n",
       "      <td>0.159574</td>\n",
       "      <td>0.223881</td>\n",
       "      <td>30.0</td>\n",
       "      <td>188.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NimR</td>\n",
       "      <td>2.691412e-07</td>\n",
       "      <td>2.691412e-07</td>\n",
       "      <td>0.375000</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CRP</td>\n",
       "      <td>1.599720e-17</td>\n",
       "      <td>1.599720e-17</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.526316</td>\n",
       "      <td>10.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>FeHm</td>\n",
       "      <td>3.036741e-21</td>\n",
       "      <td>1.214696e-20</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>0.148936</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>28.0</td>\n",
       "      <td>188.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NarP</td>\n",
       "      <td>2.127768e-24</td>\n",
       "      <td>6.383303e-24</td>\n",
       "      <td>0.435897</td>\n",
       "      <td>0.680000</td>\n",
       "      <td>0.531250</td>\n",
       "      <td>17.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>PurR</td>\n",
       "      <td>6.466229e-18</td>\n",
       "      <td>5.819606e-17</td>\n",
       "      <td>0.244898</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>0.380952</td>\n",
       "      <td>12.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MetR</td>\n",
       "      <td>3.339324e-08</td>\n",
       "      <td>2.337527e-07</td>\n",
       "      <td>0.235294</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>CRP</td>\n",
       "      <td>4.472261e-31</td>\n",
       "      <td>2.012517e-30</td>\n",
       "      <td>0.306667</td>\n",
       "      <td>0.920000</td>\n",
       "      <td>0.460000</td>\n",
       "      <td>23.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ArgR</td>\n",
       "      <td>2.056349e-16</td>\n",
       "      <td>4.112697e-16</td>\n",
       "      <td>0.538462</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.700000</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>CRP</td>\n",
       "      <td>1.210653e-07</td>\n",
       "      <td>3.026632e-07</td>\n",
       "      <td>0.225806</td>\n",
       "      <td>0.280000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>7.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   regulator        pvalue        qvalue  precision    recall   f1score    TP  \\\n",
       "0       TrpR  1.167011e-19  1.167011e-19   0.692308  0.818182  0.750000   9.0   \n",
       "7       FeHm  1.385089e-10  1.385089e-09   0.375000  0.159574  0.223881  30.0   \n",
       "9       NimR  2.691412e-07  2.691412e-07   0.375000  0.750000  0.500000   3.0   \n",
       "11       CRP  1.599720e-17  1.599720e-17   0.769231  0.400000  0.526316  10.0   \n",
       "14      FeHm  3.036741e-21  1.214696e-20   0.777778  0.148936  0.250000  28.0   \n",
       "16      NarP  2.127768e-24  6.383303e-24   0.435897  0.680000  0.531250  17.0   \n",
       "18      PurR  6.466229e-18  5.819606e-17   0.244898  0.857143  0.380952  12.0   \n",
       "19      MetR  3.339324e-08  2.337527e-07   0.235294  0.800000  0.363636   4.0   \n",
       "23       CRP  4.472261e-31  2.012517e-30   0.306667  0.920000  0.460000  23.0   \n",
       "29      ArgR  2.056349e-16  4.112697e-16   0.538462  1.000000  0.700000   7.0   \n",
       "36       CRP  1.210653e-07  3.026632e-07   0.225806  0.280000  0.250000   7.0   \n",
       "\n",
       "    regulon_size  imodulon_size  n_regs  \n",
       "0           11.0           13.0     1.0  \n",
       "7          188.0           80.0     1.0  \n",
       "9            4.0            8.0     1.0  \n",
       "11          25.0           13.0     1.0  \n",
       "14         188.0           36.0     1.0  \n",
       "16          25.0           39.0     1.0  \n",
       "18          14.0           49.0     1.0  \n",
       "19           5.0           17.0     1.0  \n",
       "23          25.0           75.0     1.0  \n",
       "29           7.0           13.0     1.0  \n",
       "36          25.0           31.0     1.0  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regulatory_imodulons = ica_data.imodulon_table[ica_data.imodulon_table.regulator.notnull()]\n",
    "print(len(ica_data.imodulon_table),'Total iModulons')\n",
    "print(len(regulatory_imodulons),'Regulatory iModulons')\n",
    "regulatory_imodulons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can rename iModulons in this jupyter notebook, or you can save the iModulon table as a CSV and edit it in Excel.\n",
    "\n",
    "If two iModulons have the same regulator (e.g. 'Reg'), they will be named 'Reg-1' and 'Reg-2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:Duplicate iModulon names detected. iModulon 7 will be renamed to FeHm-1\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 11 will be renamed to CRP-1\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 14 will be renamed to FeHm-2\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 23 will be renamed to CRP-2\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 36 will be renamed to CRP-3\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TrpR</th>\n",
       "      <td>TrpR</td>\n",
       "      <td>1.167011e-19</td>\n",
       "      <td>1.167011e-19</td>\n",
       "      <td>0.692308</td>\n",
       "      <td>0.818182</td>\n",
       "      <td>0.75</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     regulator        pvalue        qvalue  precision    recall  f1score   TP  \\\n",
       "TrpR      TrpR  1.167011e-19  1.167011e-19   0.692308  0.818182     0.75  9.0   \n",
       "1          NaN           NaN           NaN        NaN       NaN      NaN  NaN   \n",
       "2          NaN           NaN           NaN        NaN       NaN      NaN  NaN   \n",
       "3          NaN           NaN           NaN        NaN       NaN      NaN  NaN   \n",
       "4          NaN           NaN           NaN        NaN       NaN      NaN  NaN   \n",
       "\n",
       "      regulon_size  imodulon_size  n_regs  \n",
       "TrpR          11.0           13.0     1.0  \n",
       "1              NaN            NaN     NaN  \n",
       "2              NaN            NaN     NaN  \n",
       "3              NaN            NaN     NaN  \n",
       "4              NaN            NaN     NaN  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.rename_imodulons(regulatory_imodulons.regulator.to_dict())\n",
    "ica_data.imodulon_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "regulatory_imodulons = ica_data.imodulon_table[ica_data.imodulon_table.regulator.notnull()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functional iModulons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GO annotations and KEGG pathways/modules were generated in the 1_create_the_gene_table.ipynb notebook. Enrichments will be calculated in this notebook, and further curated in the 3_manual_iModulon_curation notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GO Enrichments"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First load the Gene Ontology annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_ontology</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Synpcc7942_1426</td>\n",
       "      <td>cbbL</td>\n",
       "      <td>carboxysome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Synpcc7942_1534</td>\n",
       "      <td>ago</td>\n",
       "      <td>DNA endonuclease activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Synpcc7942_1535</td>\n",
       "      <td>rbcX</td>\n",
       "      <td>cytoplasm|carboxysome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Synpcc7942_1891</td>\n",
       "      <td>labA</td>\n",
       "      <td>circadian rhythm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Synpcc7942_2462</td>\n",
       "      <td>PSK</td>\n",
       "      <td>ATP binding|D-ribulokinase activity</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id gene_name                        gene_ontology\n",
       "0  Synpcc7942_1426      cbbL                          carboxysome\n",
       "1  Synpcc7942_1534       ago            DNA endonuclease activity\n",
       "2  Synpcc7942_1535      rbcX                cytoplasm|carboxysome\n",
       "3  Synpcc7942_1891      labA                     circadian rhythm\n",
       "4  Synpcc7942_2462       PSK  ATP binding|D-ribulokinase activity"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_GO = pd.read_csv(path.join(external_data,'GO_annotations_curated.csv'),index_col=0)\n",
    "DF_GO.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_GO_enrich = ica_data.compute_annotation_enrichment(DF_GO,'gene_ontology')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>gene_ontology</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>17</td>\n",
       "      <td>entrainment of circadian clock</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>17</td>\n",
       "      <td>identical protein binding</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>17</td>\n",
       "      <td>protein binding</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>0.000542</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>0.307692</td>\n",
       "      <td>2.0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>17</td>\n",
       "      <td>identical protein binding|protein binding|entr...</td>\n",
       "      <td>0.001124</td>\n",
       "      <td>0.005370</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>17</td>\n",
       "      <td>identical protein binding|protein binding|entr...</td>\n",
       "      <td>0.001124</td>\n",
       "      <td>0.005370</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   imodulon                                      gene_ontology    pvalue  \\\n",
       "0        17                     entrainment of circadian clock  0.000003   \n",
       "1        17                          identical protein binding  0.000008   \n",
       "2        17                                    protein binding  0.000038   \n",
       "3        17  identical protein binding|protein binding|entr...  0.001124   \n",
       "4        17  identical protein binding|protein binding|entr...  0.001124   \n",
       "\n",
       "     qvalue  precision    recall   f1score   TP  target_set_size  \\\n",
       "0  0.000109   0.666667  0.666667  0.666667  2.0              3.0   \n",
       "1  0.000181   0.666667  0.400000  0.500000  2.0             15.0   \n",
       "2  0.000542   0.666667  0.200000  0.307692  2.0             63.0   \n",
       "3  0.005370   0.333333  1.000000  0.500000  1.0              1.0   \n",
       "4  0.005370   0.333333  1.000000  0.500000  1.0              1.0   \n",
       "\n",
       "   imodulon_size  \n",
       "0            3.0  \n",
       "1            3.0  \n",
       "2            3.0  \n",
       "3            3.0  \n",
       "4            3.0  "
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_GO_enrich.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## KEGG Enrichments"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load KEGG mapping\n",
    "The `kegg_mapping.csv` file contains KEGG orthologies, pathways, modules, and reactions. Only pathways and modules are relevant to iModulon characterization."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['KEGG_pathway' 'KEGG_module' 'KEGG_reaction']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>database</th>\n",
       "      <th>kegg_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Synpcc7942_0001</td>\n",
       "      <td>KEGG_pathway</td>\n",
       "      <td>map00230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Synpcc7942_0001</td>\n",
       "      <td>KEGG_pathway</td>\n",
       "      <td>map00240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Synpcc7942_0001</td>\n",
       "      <td>KEGG_pathway</td>\n",
       "      <td>map01100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Synpcc7942_0001</td>\n",
       "      <td>KEGG_pathway</td>\n",
       "      <td>map03030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Synpcc7942_0001</td>\n",
       "      <td>KEGG_pathway</td>\n",
       "      <td>map03430</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene_id      database   kegg_id\n",
       "6   Synpcc7942_0001  KEGG_pathway  map00230\n",
       "7   Synpcc7942_0001  KEGG_pathway  map00240\n",
       "8   Synpcc7942_0001  KEGG_pathway  map01100\n",
       "9   Synpcc7942_0001  KEGG_pathway  map03030\n",
       "10  Synpcc7942_0001  KEGG_pathway  map03430"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_KEGG = pd.read_csv(path.join(external_data,'kegg_mapping.csv'),index_col=0)\n",
    "print(DF_KEGG.database.unique())\n",
    "DF_KEGG.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "kegg_pathways = DF_KEGG[DF_KEGG.database == 'KEGG_pathway']\n",
    "kegg_modules = DF_KEGG[DF_KEGG.database == 'KEGG_module']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Perform enrichment\n",
    "Uses the `compute_annotation_enrichment` function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_pathway_enrich = ica_data.compute_annotation_enrichment(kegg_pathways,'kegg_id')\n",
    "DF_module_enrich = ica_data.compute_annotation_enrichment(kegg_modules,'kegg_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>map00920</td>\n",
       "      <td>2.868428e-10</td>\n",
       "      <td>5.880278e-08</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.370370</td>\n",
       "      <td>5.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>map02010</td>\n",
       "      <td>1.062695e-05</td>\n",
       "      <td>1.089262e-03</td>\n",
       "      <td>0.571429</td>\n",
       "      <td>0.061538</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>4.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>map02020</td>\n",
       "      <td>4.145330e-07</td>\n",
       "      <td>4.248963e-05</td>\n",
       "      <td>0.454545</td>\n",
       "      <td>0.113636</td>\n",
       "      <td>0.181818</td>\n",
       "      <td>5.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>map02025</td>\n",
       "      <td>3.622367e-07</td>\n",
       "      <td>4.248963e-05</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>0.235294</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>4.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6</td>\n",
       "      <td>map00190</td>\n",
       "      <td>1.486547e-10</td>\n",
       "      <td>3.047421e-08</td>\n",
       "      <td>0.346154</td>\n",
       "      <td>0.195652</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>9.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   imodulon   kegg_id        pvalue        qvalue  precision    recall  \\\n",
       "0         1  map00920  2.868428e-10  5.880278e-08   0.714286  0.250000   \n",
       "1         1  map02010  1.062695e-05  1.089262e-03   0.571429  0.061538   \n",
       "2         4  map02020  4.145330e-07  4.248963e-05   0.454545  0.113636   \n",
       "3         4  map02025  3.622367e-07  4.248963e-05   0.363636  0.235294   \n",
       "4         6  map00190  1.486547e-10  3.047421e-08   0.346154  0.195652   \n",
       "\n",
       "    f1score   TP  target_set_size  imodulon_size  \n",
       "0  0.370370  5.0             20.0            7.0  \n",
       "1  0.111111  4.0             65.0            7.0  \n",
       "2  0.181818  5.0             44.0           11.0  \n",
       "3  0.285714  4.0             17.0           11.0  \n",
       "4  0.250000  9.0             46.0           26.0  "
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_pathway_enrich.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>M00185</td>\n",
       "      <td>1.157162e-09</td>\n",
       "      <td>2.360609e-07</td>\n",
       "      <td>0.571429</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.533333</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>M00190</td>\n",
       "      <td>2.522720e-05</td>\n",
       "      <td>5.146350e-03</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>M00507</td>\n",
       "      <td>1.035552e-06</td>\n",
       "      <td>2.112526e-04</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.352941</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6</td>\n",
       "      <td>M00145</td>\n",
       "      <td>7.718266e-15</td>\n",
       "      <td>1.574526e-12</td>\n",
       "      <td>0.346154</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.409091</td>\n",
       "      <td>9.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>M00155</td>\n",
       "      <td>3.061383e-07</td>\n",
       "      <td>6.245221e-05</td>\n",
       "      <td>0.157895</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>19.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   imodulon kegg_id        pvalue        qvalue  precision  recall   f1score  \\\n",
       "0         1  M00185  1.157162e-09  2.360609e-07   0.571429     0.5  0.533333   \n",
       "1         3  M00190  2.522720e-05  5.146350e-03   0.333333     0.5  0.400000   \n",
       "2         4  M00507  1.035552e-06  2.112526e-04   0.272727     0.5  0.352941   \n",
       "3         6  M00145  7.718266e-15  1.574526e-12   0.346154     0.5  0.409091   \n",
       "4        13  M00155  3.061383e-07  6.245221e-05   0.157895     1.0  0.272727   \n",
       "\n",
       "    TP  target_set_size  imodulon_size  \n",
       "0  4.0              8.0            7.0  \n",
       "1  2.0              4.0            6.0  \n",
       "2  3.0              6.0           11.0  \n",
       "3  9.0             18.0           26.0  \n",
       "4  3.0              3.0           19.0  "
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_module_enrich.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert KEGG IDs to human-readable names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "32797f870b554d27817b3adbad97aaf7",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(HTML(value=''), FloatProgress(value=0.0, max=26.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5d1dee5d210c40448f78f9bdac82bf8f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(HTML(value=''), FloatProgress(value=0.0, max=20.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "for idx,key in tqdm(DF_pathway_enrich.kegg_id.items(),total=len(DF_pathway_enrich)):\n",
    "    text = REST.kegg_find('pathway',key).read()\n",
    "    try:\n",
    "        name = re.search('\\t(.*)\\n',text).group(1)\n",
    "        DF_pathway_enrich.loc[idx,'pathway_name'] = name\n",
    "    except AttributeError:\n",
    "        DF_pathway_enrich.loc[idx,'pathway_name'] = None\n",
    "    \n",
    "for idx,key in tqdm(DF_module_enrich.kegg_id.items(),total=len(DF_module_enrich)):\n",
    "    text = REST.kegg_find('module',key).read()\n",
    "    try:\n",
    "        name = re.search('\\t(.*)\\n',text).group(1)\n",
    "        DF_module_enrich.loc[idx,'module_name'] = name\n",
    "    except AttributeError:\n",
    "        DF_module_enrich.loc[idx,'module_name'] = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>pathway_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>map00920</td>\n",
       "      <td>2.868428e-10</td>\n",
       "      <td>5.880278e-08</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.370370</td>\n",
       "      <td>5.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Sulfur metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>map02010</td>\n",
       "      <td>1.062695e-05</td>\n",
       "      <td>1.089262e-03</td>\n",
       "      <td>0.571429</td>\n",
       "      <td>0.061538</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>4.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ABC transporters</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>map02020</td>\n",
       "      <td>4.145330e-07</td>\n",
       "      <td>4.248963e-05</td>\n",
       "      <td>0.454545</td>\n",
       "      <td>0.113636</td>\n",
       "      <td>0.181818</td>\n",
       "      <td>5.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>Two-component system</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>map02025</td>\n",
       "      <td>3.622367e-07</td>\n",
       "      <td>4.248963e-05</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>0.235294</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>4.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>Biofilm formation - Pseudomonas aeruginosa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6</td>\n",
       "      <td>map00190</td>\n",
       "      <td>1.486547e-10</td>\n",
       "      <td>3.047421e-08</td>\n",
       "      <td>0.346154</td>\n",
       "      <td>0.195652</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>9.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   imodulon   kegg_id        pvalue        qvalue  precision    recall  \\\n",
       "0         1  map00920  2.868428e-10  5.880278e-08   0.714286  0.250000   \n",
       "1         1  map02010  1.062695e-05  1.089262e-03   0.571429  0.061538   \n",
       "2         4  map02020  4.145330e-07  4.248963e-05   0.454545  0.113636   \n",
       "3         4  map02025  3.622367e-07  4.248963e-05   0.363636  0.235294   \n",
       "4         6  map00190  1.486547e-10  3.047421e-08   0.346154  0.195652   \n",
       "\n",
       "    f1score   TP  target_set_size  imodulon_size  \\\n",
       "0  0.370370  5.0             20.0            7.0   \n",
       "1  0.111111  4.0             65.0            7.0   \n",
       "2  0.181818  5.0             44.0           11.0   \n",
       "3  0.285714  4.0             17.0           11.0   \n",
       "4  0.250000  9.0             46.0           26.0   \n",
       "\n",
       "                                 pathway_name  \n",
       "0                           Sulfur metabolism  \n",
       "1                            ABC transporters  \n",
       "2                        Two-component system  \n",
       "3  Biofilm formation - Pseudomonas aeruginosa  \n",
       "4                   Oxidative phosphorylation  "
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_pathway_enrich.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>module_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>M00185</td>\n",
       "      <td>1.157162e-09</td>\n",
       "      <td>2.360609e-07</td>\n",
       "      <td>0.571429</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.533333</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>M00190</td>\n",
       "      <td>2.522720e-05</td>\n",
       "      <td>5.146350e-03</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>M00507</td>\n",
       "      <td>1.035552e-06</td>\n",
       "      <td>2.112526e-04</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.352941</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6</td>\n",
       "      <td>M00145</td>\n",
       "      <td>7.718266e-15</td>\n",
       "      <td>1.574526e-12</td>\n",
       "      <td>0.346154</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.409091</td>\n",
       "      <td>9.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>NAD(P)H:quinone oxidoreductase, chloroplasts a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>M00155</td>\n",
       "      <td>3.061383e-07</td>\n",
       "      <td>6.245221e-05</td>\n",
       "      <td>0.157895</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>Cytochrome c oxidase, prokaryotes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   imodulon kegg_id        pvalue        qvalue  precision  recall   f1score  \\\n",
       "0         1  M00185  1.157162e-09  2.360609e-07   0.571429     0.5  0.533333   \n",
       "1         3  M00190  2.522720e-05  5.146350e-03   0.333333     0.5  0.400000   \n",
       "2         4  M00507  1.035552e-06  2.112526e-04   0.272727     0.5  0.352941   \n",
       "3         6  M00145  7.718266e-15  1.574526e-12   0.346154     0.5  0.409091   \n",
       "4        13  M00155  3.061383e-07  6.245221e-05   0.157895     1.0  0.272727   \n",
       "\n",
       "    TP  target_set_size  imodulon_size  \\\n",
       "0  4.0              8.0            7.0   \n",
       "1  2.0              4.0            6.0   \n",
       "2  3.0              6.0           11.0   \n",
       "3  9.0             18.0           26.0   \n",
       "4  3.0              3.0           19.0   \n",
       "\n",
       "                                         module_name  \n",
       "0                                                NaN  \n",
       "1                                                NaN  \n",
       "2                                                NaN  \n",
       "3  NAD(P)H:quinone oxidoreductase, chloroplasts a...  \n",
       "4                  Cytochrome c oxidase, prokaryotes  "
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_module_enrich.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_GO_enrich['source'] = 'GO'\n",
    "DF_pathway_enrich['source'] = 'KEGG pathways'\n",
    "DF_module_enrich['source'] = 'KEGG modules'\n",
    "\n",
    "DF_GO_enrich.rename({'gene_ontology':'annotation'},axis=1, inplace=True)\n",
    "DF_pathway_enrich.rename({'kegg_id':'annotation'},axis=1, inplace=True)\n",
    "DF_module_enrich.rename({'kegg_id':'annotation'},axis=1, inplace=True)\n",
    "\n",
    "DF_enrichments = pd.concat([DF_GO_enrich, DF_pathway_enrich, DF_module_enrich])\n",
    "DF_enrichments.to_csv(path.join(data_dir,'functional_enrichments.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check for single gene iModulons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some iModulons are dominated by a single, high-coefficient gene. These iModulons may result from:\n",
    "1. Overdecomposition of the dataset to identify noisy genes\n",
    "1. Artificial knock-out of single genes\n",
    "1. Regulons with only one target gene\n",
    "\n",
    "No matter what causes these iModulons, it is important to be aware of them. The find_single_gene_imodulons function identifies iModulons that are likely dominated by a single gene.\n",
    "\n",
    "The iModulons identified by ``find_single_gene_imodulons`` may contain more than one gene, since a threshold-agnostic method is used to identify these iModulons."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg_imods = ica_data.find_single_gene_imodulons(save=True)\n",
    "len(sg_imods)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i,mod in enumerate(sg_imods):\n",
    "    ica_data.rename_imodulons({mod:'SG_'+str(i+1)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>single_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SG_1</th>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_2</th>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_3</th>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_4</th>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_5</th>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_6</th>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_7</th>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_8</th>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     single_gene\n",
       "SG_1        True\n",
       "SG_2        True\n",
       "SG_3        True\n",
       "SG_4        True\n",
       "SG_5        True\n",
       "SG_6        True\n",
       "SG_7        True\n",
       "SG_8        True"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.imodulon_table[ica_data.imodulon_table.single_gene == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_weight</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>accession</th>\n",
       "      <th>old_locus_tag</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_product</th>\n",
       "      <th>COG</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>operon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Synpcc7942_B2641</th>\n",
       "      <td>0.601549</td>\n",
       "      <td>Synpcc7942_B2641</td>\n",
       "      <td>CP000101.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>36267</td>\n",
       "      <td>36392</td>\n",
       "      <td>+</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>No COG annotation</td>\n",
       "      <td>Q31JT7</td>\n",
       "      <td>Op60</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gene_weight         gene_name   accession old_locus_tag  \\\n",
       "Synpcc7942_B2641     0.601549  Synpcc7942_B2641  CP000101.1           NaN   \n",
       "\n",
       "                  start    end strand          gene_product  \\\n",
       "Synpcc7942_B2641  36267  36392      +  hypothetical protein   \n",
       "\n",
       "                                COG uniprot operon  \n",
       "Synpcc7942_B2641  No COG annotation  Q31JT7   Op60  "
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.view_imodulon('SG_1')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Old characterizations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon.io import *\n",
    "from pymodulon.compare import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [],
   "source": [
    "old = load_json_model('../Tahani/selon.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>subject</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Synpcc7942_R0020</td>\n",
       "      <td>SYNPCC7942_RS12910</td>\n",
       "      <td>tRNA-Asp1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Synpcc7942_R0033</td>\n",
       "      <td>SYNPCC7942_RS02990</td>\n",
       "      <td>tRNA-Ile1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Synpcc7942_R0016</td>\n",
       "      <td>SYNPCC7942_RS12495</td>\n",
       "      <td>tRNA-Ser2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Synpcc7942_R0048</td>\n",
       "      <td>SYNPCC7942_RS07055</td>\n",
       "      <td>tRNA-Leu3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Synpcc7942_R0030</td>\n",
       "      <td>SYNPCC7942_RS01480</td>\n",
       "      <td>tRNA-Ala3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene             subject       name\n",
       "0  Synpcc7942_R0020  SYNPCC7942_RS12910  tRNA-Asp1\n",
       "1  Synpcc7942_R0033  SYNPCC7942_RS02990  tRNA-Ile1\n",
       "2  Synpcc7942_R0016  SYNPCC7942_RS12495  tRNA-Ser2\n",
       "3  Synpcc7942_R0048  SYNPCC7942_RS07055  tRNA-Leu3\n",
       "4  Synpcc7942_R0030  SYNPCC7942_RS01480  tRNA-Ala3"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping = pd.read_csv('../data/external/1_gene_mapping_biocyc.csv')\n",
    "mapping = mapping.rename(columns={'objecct_id':'gene','new_locus':'subject'})\n",
    "mapping['gene'] = mapping['gene'].str.replace('SYNPCC7942', 'Synpcc7942')\n",
    "mapping.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.43.0 (0)\n",
       " -->\n",
       "<!-- Title: %3 Pages: 1 -->\n",
       "<svg width=\"720pt\" height=\"569pt\"\n",
       " viewBox=\"0.00 0.00 720.00 569.43\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1.1 1.1) rotate(0) translate(4 620)\">\n",
       "<title>%3</title>\n",
       "<polygon fill=\"white\" stroke=\"transparent\" points=\"-4,4 -4,-620 785,-620 785,4 -4,4\"/>\n",
       "<!-- data2_Biofilm&#45;1 -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>data2_Biofilm&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"39.5\" y=\"-535.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Biofilm&#45;1</text>\n",
       "</g>\n",
       "<!-- data2_Biofilm&#45;2 -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>data2_Biofilm&#45;2</title>\n",
       "<text text-anchor=\"middle\" x=\"206.5\" y=\"-535.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Biofilm&#45;2</text>\n",
       "</g>\n",
       "<!-- data2_CMM -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>data2_CMM</title>\n",
       "<text text-anchor=\"middle\" x=\"331.5\" y=\"-535.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">CMM</text>\n",
       "</g>\n",
       "<!-- data2_Carbon Transport -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>data2_Carbon Transport</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-542.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Carbon T&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-527.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ransport</text>\n",
       "</g>\n",
       "<!-- data2_CmpR -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>data2_CmpR</title>\n",
       "<text text-anchor=\"middle\" x=\"526.5\" y=\"-535.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">CmpR</text>\n",
       "</g>\n",
       "<!-- data2_Comp&#45;1 -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>data2_Comp&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"624.5\" y=\"-535.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Comp&#45;1</text>\n",
       "</g>\n",
       "<!-- data2_Comp&#45;2 -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>data2_Comp&#45;2</title>\n",
       "<text text-anchor=\"middle\" x=\"730.5\" y=\"-535.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Comp&#45;2</text>\n",
       "</g>\n",
       "<!-- data2_CysR -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>data2_CysR</title>\n",
       "<text text-anchor=\"middle\" x=\"67.5\" y=\"-431.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">CysR</text>\n",
       "</g>\n",
       "<!-- data2_Cytc Oxidases  -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>data2_Cytc Oxidases </title>\n",
       "<text text-anchor=\"middle\" x=\"206.5\" y=\"-438.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Cytc Ox&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"206.5\" y=\"-423.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">idases </text>\n",
       "</g>\n",
       "<!-- data2_Dark&#45;Induced Transition -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>data2_Dark&#45;Induced Transition</title>\n",
       "<text text-anchor=\"middle\" x=\"331.5\" y=\"-438.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Dark&#45;Induce&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"331.5\" y=\"-423.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">d Transition</text>\n",
       "</g>\n",
       "<!-- data2_Fur&#45;Like -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>data2_Fur&#45;Like</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-431.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Fur&#45;Like</text>\n",
       "</g>\n",
       "<!-- data2_HSR -->\n",
       "<g id=\"node12\" class=\"node\">\n",
       "<title>data2_HSR</title>\n",
       "<text text-anchor=\"middle\" x=\"526.5\" y=\"-431.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">HSR</text>\n",
       "</g>\n",
       "<!-- data2_IdiB -->\n",
       "<g id=\"node13\" class=\"node\">\n",
       "<title>data2_IdiB</title>\n",
       "<text text-anchor=\"middle\" x=\"624.5\" y=\"-431.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">IdiB</text>\n",
       "</g>\n",
       "<!-- data2_Membrane Biogenesis -->\n",
       "<g id=\"node14\" class=\"node\">\n",
       "<title>data2_Membrane Biogenesis</title>\n",
       "<text text-anchor=\"middle\" x=\"730.5\" y=\"-438.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Membrane &#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"730.5\" y=\"-423.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Biogenesis</text>\n",
       "</g>\n",
       "<!-- data2_NblS&#45;RpaB&#45;SrrA&#45;2  -->\n",
       "<g id=\"node15\" class=\"node\">\n",
       "<title>data2_NblS&#45;RpaB&#45;SrrA&#45;2 </title>\n",
       "<text text-anchor=\"middle\" x=\"67.5\" y=\"-334.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">NblS&#45;Rpa&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"67.5\" y=\"-319.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">B&#45;SrrA&#45;2 </text>\n",
       "</g>\n",
       "<!-- data2_NtcA&#45;1 -->\n",
       "<g id=\"node16\" class=\"node\">\n",
       "<title>data2_NtcA&#45;1</title>\n",
       "<text text-anchor=\"middle\" x=\"206.5\" y=\"-327.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">NtcA&#45;1</text>\n",
       "</g>\n",
       "<!-- data2_NtcA&#45;2 -->\n",
       "<g id=\"node17\" class=\"node\">\n",
       "<title>data2_NtcA&#45;2</title>\n",
       "<text text-anchor=\"middle\" x=\"331.5\" y=\"-327.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">NtcA&#45;2</text>\n",
       "</g>\n",
       "<!-- data2_NtcA&#45;3 -->\n",
       "<g id=\"node18\" class=\"node\">\n",
       "<title>data2_NtcA&#45;3</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-327.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">NtcA&#45;3</text>\n",
       "</g>\n",
       "<!-- data2_NtcA&#45;4 -->\n",
       "<g id=\"node19\" class=\"node\">\n",
       "<title>data2_NtcA&#45;4</title>\n",
       "<text text-anchor=\"middle\" x=\"526.5\" y=\"-327.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">NtcA&#45;4</text>\n",
       "</g>\n",
       "<!-- data2_OxPh NADPH ** -->\n",
       "<g id=\"node20\" class=\"node\">\n",
       "<title>data2_OxPh NADPH **</title>\n",
       "<text text-anchor=\"middle\" x=\"624.5\" y=\"-334.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">OxPh N&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"624.5\" y=\"-319.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ADPH **</text>\n",
       "</g>\n",
       "<!-- data2_OxiTol -->\n",
       "<g id=\"node21\" class=\"node\">\n",
       "<title>data2_OxiTol</title>\n",
       "<text text-anchor=\"middle\" x=\"730.5\" y=\"-327.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">OxiTol</text>\n",
       "</g>\n",
       "<!-- data2_PS -->\n",
       "<g id=\"node22\" class=\"node\">\n",
       "<title>data2_PS</title>\n",
       "<text text-anchor=\"middle\" x=\"67.5\" y=\"-223.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">PS</text>\n",
       "</g>\n",
       "<!-- data2_PSII&#45;Cofactors -->\n",
       "<g id=\"node23\" class=\"node\">\n",
       "<title>data2_PSII&#45;Cofactors</title>\n",
       "<text text-anchor=\"middle\" x=\"178.5\" y=\"-230.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">PSII&#45;Co&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"178.5\" y=\"-215.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">factors</text>\n",
       "</g>\n",
       "<!-- data2_Phage Related -->\n",
       "<g id=\"node24\" class=\"node\">\n",
       "<title>data2_Phage Related</title>\n",
       "<text text-anchor=\"middle\" x=\"331.5\" y=\"-230.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Phage &#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"331.5\" y=\"-215.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Related</text>\n",
       "</g>\n",
       "<!-- data2_PhoTax -->\n",
       "<g id=\"node25\" class=\"node\">\n",
       "<title>data2_PhoTax</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-223.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">PhoTax</text>\n",
       "</g>\n",
       "<!-- data2_Purine Metabolism -->\n",
       "<g id=\"node26\" class=\"node\">\n",
       "<title>data2_Purine Metabolism</title>\n",
       "<text text-anchor=\"middle\" x=\"526.5\" y=\"-230.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Purine M&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"526.5\" y=\"-215.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">etabolism</text>\n",
       "</g>\n",
       "<!-- data2_RpaA/RpaB -->\n",
       "<g id=\"node27\" class=\"node\">\n",
       "<title>data2_RpaA/RpaB</title>\n",
       "<text text-anchor=\"middle\" x=\"624.5\" y=\"-223.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">RpaA/RpaB</text>\n",
       "</g>\n",
       "<!-- data2_RpaB -->\n",
       "<g id=\"node28\" class=\"node\">\n",
       "<title>data2_RpaB</title>\n",
       "<text text-anchor=\"middle\" x=\"730.5\" y=\"-223.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">RpaB</text>\n",
       "</g>\n",
       "<!-- data2_SG_1 -->\n",
       "<g id=\"node29\" class=\"node\">\n",
       "<title>data2_SG_1</title>\n",
       "<text text-anchor=\"middle\" x=\"67.5\" y=\"-119.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">SG_1</text>\n",
       "</g>\n",
       "<!-- data2_SG_3 -->\n",
       "<g id=\"node30\" class=\"node\">\n",
       "<title>data2_SG_3</title>\n",
       "<text text-anchor=\"middle\" x=\"206.5\" y=\"-119.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">SG_3</text>\n",
       "</g>\n",
       "<!-- data2_SG_4 -->\n",
       "<g id=\"node31\" class=\"node\">\n",
       "<title>data2_SG_4</title>\n",
       "<text text-anchor=\"middle\" x=\"331.5\" y=\"-119.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">SG_4</text>\n",
       "</g>\n",
       "<!-- data2_SG_8 -->\n",
       "<g id=\"node32\" class=\"node\">\n",
       "<title>data2_SG_8</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-119.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">SG_8</text>\n",
       "</g>\n",
       "<!-- data2_STR -->\n",
       "<g id=\"node33\" class=\"node\">\n",
       "<title>data2_STR</title>\n",
       "<text text-anchor=\"middle\" x=\"526.5\" y=\"-119.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">STR</text>\n",
       "</g>\n",
       "<!-- data2_SphR -->\n",
       "<g id=\"node34\" class=\"node\">\n",
       "<title>data2_SphR</title>\n",
       "<text text-anchor=\"middle\" x=\"107.5\" y=\"-535.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">SphR</text>\n",
       "</g>\n",
       "<!-- data2_State Transition -->\n",
       "<g id=\"node35\" class=\"node\">\n",
       "<title>data2_State Transition</title>\n",
       "<text text-anchor=\"middle\" x=\"624.5\" y=\"-126.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">State Tr&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"624.5\" y=\"-111.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">ansition</text>\n",
       "</g>\n",
       "<!-- data2_Translation -->\n",
       "<g id=\"node36\" class=\"node\">\n",
       "<title>data2_Translation</title>\n",
       "<text text-anchor=\"middle\" x=\"730.5\" y=\"-126.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">Trans&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"730.5\" y=\"-111.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">lation</text>\n",
       "</g>\n",
       "<!-- data2_U_11 -->\n",
       "<g id=\"node37\" class=\"node\">\n",
       "<title>data2_U_11</title>\n",
       "<text text-anchor=\"middle\" x=\"39.5\" y=\"-15.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">U_11</text>\n",
       "</g>\n",
       "<!-- data2_U_12 -->\n",
       "<g id=\"node38\" class=\"node\">\n",
       "<title>data2_U_12</title>\n",
       "<text text-anchor=\"middle\" x=\"94.5\" y=\"-15.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">U_12</text>\n",
       "</g>\n",
       "<!-- data2_U_5 -->\n",
       "<g id=\"node39\" class=\"node\">\n",
       "<title>data2_U_5</title>\n",
       "<text text-anchor=\"middle\" x=\"242.5\" y=\"-223.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">U_5</text>\n",
       "</g>\n",
       "<!-- data2_U_7 -->\n",
       "<g id=\"node40\" class=\"node\">\n",
       "<title>data2_U_7</title>\n",
       "<text text-anchor=\"middle\" x=\"206.5\" y=\"-15.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">U_7</text>\n",
       "</g>\n",
       "<!-- data2_U_9 -->\n",
       "<g id=\"node41\" class=\"node\">\n",
       "<title>data2_U_9</title>\n",
       "<text text-anchor=\"middle\" x=\"331.5\" y=\"-15.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">U_9</text>\n",
       "</g>\n",
       "<!-- data2_rpaA/rpopD6  -->\n",
       "<g id=\"node42\" class=\"node\">\n",
       "<title>data2_rpaA/rpopD6 </title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-22.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">rpaA/r&#45;</text>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-7.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">popD6 </text>\n",
       "</g>\n",
       "<!-- data1_0 -->\n",
       "<g id=\"node43\" class=\"node\">\n",
       "<title>data1_0</title>\n",
       "<text text-anchor=\"middle\" x=\"206.5\" y=\"-73.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">0</text>\n",
       "</g>\n",
       "<!-- data1_0&#45;&gt;data2_U_7 -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>data1_0&#45;&gt;data2_U_7</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.12\" d=\"M206.5,-58.89C206.5,-53.57 206.5,-47.63 206.5,-42.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.12\" points=\"208.3,-42 206.5,-37 204.7,-42 208.3,-42\"/>\n",
       "</g>\n",
       "<!-- data1_1 -->\n",
       "<g id=\"node44\" class=\"node\">\n",
       "<title>data1_1</title>\n",
       "<text text-anchor=\"middle\" x=\"67.5\" y=\"-489.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">1</text>\n",
       "</g>\n",
       "<!-- data1_1&#45;&gt;data2_CysR -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>data1_1&#45;&gt;data2_CysR</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.04\" d=\"M67.5,-474.89C67.5,-469.57 67.5,-463.63 67.5,-458.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.04\" points=\"69.27,-458 67.5,-453 65.73,-458 69.27,-458\"/>\n",
       "</g>\n",
       "<!-- data1_10 -->\n",
       "<g id=\"node45\" class=\"node\">\n",
       "<title>data1_10</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-385.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">10</text>\n",
       "</g>\n",
       "<!-- data1_10&#45;&gt;data2_NtcA&#45;3 -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>data1_10&#45;&gt;data2_NtcA&#45;3</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.07\" d=\"M434.5,-370.89C434.5,-365.57 434.5,-359.63 434.5,-354.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.07\" points=\"436.25,-354 434.5,-349 432.75,-354 436.25,-354\"/>\n",
       "</g>\n",
       "<!-- data1_12 -->\n",
       "<g id=\"node46\" class=\"node\">\n",
       "<title>data1_12</title>\n",
       "<text text-anchor=\"middle\" x=\"331.5\" y=\"-385.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">12</text>\n",
       "</g>\n",
       "<!-- data1_12&#45;&gt;data2_NtcA&#45;2 -->\n",
       "<g id=\"edge14\" class=\"edge\">\n",
       "<title>data1_12&#45;&gt;data2_NtcA&#45;2</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.23\" d=\"M331.5,-370.89C331.5,-365.57 331.5,-359.63 331.5,-354.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.23\" points=\"333.35,-354 331.5,-349 329.65,-354 333.35,-354\"/>\n",
       "</g>\n",
       "<!-- data1_13 -->\n",
       "<g id=\"node47\" class=\"node\">\n",
       "<title>data1_13</title>\n",
       "<text text-anchor=\"middle\" x=\"206.5\" y=\"-490.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">13</text>\n",
       "</g>\n",
       "<!-- data1_13&#45;&gt;data2_Cytc Oxidases  -->\n",
       "<g id=\"edge15\" class=\"edge\">\n",
       "<title>data1_13&#45;&gt;data2_Cytc Oxidases </title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.73\" d=\"M206.5,-475.9C206.5,-470.68 206.5,-464.84 206.5,-459.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.73\" points=\"208.25,-459.28 206.5,-454.28 204.75,-459.28 208.25,-459.28\"/>\n",
       "</g>\n",
       "<!-- data1_16 -->\n",
       "<g id=\"node48\" class=\"node\">\n",
       "<title>data1_16</title>\n",
       "<text text-anchor=\"middle\" x=\"331.5\" y=\"-73.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">16</text>\n",
       "</g>\n",
       "<!-- data1_16&#45;&gt;data2_U_9 -->\n",
       "<g id=\"edge17\" class=\"edge\">\n",
       "<title>data1_16&#45;&gt;data2_U_9</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.46\" d=\"M331.5,-58.89C331.5,-53.57 331.5,-47.63 331.5,-42.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.46\" points=\"333.45,-42 331.5,-37 329.55,-42 333.45,-42\"/>\n",
       "</g>\n",
       "<!-- data1_17 -->\n",
       "<g id=\"node49\" class=\"node\">\n",
       "<title>data1_17</title>\n",
       "<text text-anchor=\"middle\" x=\"730.5\" y=\"-281.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">17</text>\n",
       "</g>\n",
       "<!-- data1_17&#45;&gt;data2_RpaB -->\n",
       "<g id=\"edge18\" class=\"edge\">\n",
       "<title>data1_17&#45;&gt;data2_RpaB</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.43\" d=\"M730.5,-266.89C730.5,-261.57 730.5,-255.63 730.5,-250.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.43\" points=\"732.25,-250 730.5,-245 728.75,-250 732.25,-250\"/>\n",
       "</g>\n",
       "<!-- data1_18 -->\n",
       "<g id=\"node50\" class=\"node\">\n",
       "<title>data1_18</title>\n",
       "<text text-anchor=\"middle\" x=\"624.5\" y=\"-178.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">18</text>\n",
       "</g>\n",
       "<!-- data1_18&#45;&gt;data2_State Transition -->\n",
       "<g id=\"edge19\" class=\"edge\">\n",
       "<title>data1_18&#45;&gt;data2_State Transition</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.09\" d=\"M624.5,-163.9C624.5,-158.68 624.5,-152.84 624.5,-147.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.09\" points=\"626.29,-147.28 624.5,-142.28 622.71,-147.28 626.29,-147.28\"/>\n",
       "</g>\n",
       "<!-- data1_19 -->\n",
       "<g id=\"node51\" class=\"node\">\n",
       "<title>data1_19</title>\n",
       "<text text-anchor=\"middle\" x=\"206.5\" y=\"-385.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">19</text>\n",
       "</g>\n",
       "<!-- data1_19&#45;&gt;data2_NtcA&#45;1 -->\n",
       "<g id=\"edge20\" class=\"edge\">\n",
       "<title>data1_19&#45;&gt;data2_NtcA&#45;1</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.11\" d=\"M206.5,-370.89C206.5,-365.57 206.5,-359.63 206.5,-354.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.11\" points=\"208.3,-354 206.5,-349 204.7,-354 208.3,-354\"/>\n",
       "</g>\n",
       "<!-- data1_2 -->\n",
       "<g id=\"node52\" class=\"node\">\n",
       "<title>data1_2</title>\n",
       "<text text-anchor=\"middle\" x=\"331.5\" y=\"-490.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">2</text>\n",
       "</g>\n",
       "<!-- data1_2&#45;&gt;data2_Dark&#45;Induced Transition -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>data1_2&#45;&gt;data2_Dark&#45;Induced Transition</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.55\" d=\"M331.5,-475.9C331.5,-470.68 331.5,-464.84 331.5,-459.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.55\" points=\"333.49,-459.28 331.5,-454.28 329.51,-459.28 333.49,-459.28\"/>\n",
       "</g>\n",
       "<!-- data1_20 -->\n",
       "<g id=\"node53\" class=\"node\">\n",
       "<title>data1_20</title>\n",
       "<text text-anchor=\"middle\" x=\"624.5\" y=\"-489.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">20</text>\n",
       "</g>\n",
       "<!-- data1_20&#45;&gt;data2_IdiB -->\n",
       "<g id=\"edge21\" class=\"edge\">\n",
       "<title>data1_20&#45;&gt;data2_IdiB</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.17\" d=\"M624.5,-474.89C624.5,-469.57 624.5,-463.63 624.5,-458.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.17\" points=\"626.25,-458 624.5,-453 622.75,-458 626.25,-458\"/>\n",
       "</g>\n",
       "<!-- data1_22 -->\n",
       "<g id=\"node54\" class=\"node\">\n",
       "<title>data1_22</title>\n",
       "<text text-anchor=\"middle\" x=\"39.5\" y=\"-593.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">22</text>\n",
       "</g>\n",
       "<!-- data1_22&#45;&gt;data2_Biofilm&#45;1 -->\n",
       "<g id=\"edge22\" class=\"edge\">\n",
       "<title>data1_22&#45;&gt;data2_Biofilm&#45;1</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2.55\" d=\"M39.5,-578.89C39.5,-573.57 39.5,-567.63 39.5,-562.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2.55\" points=\"41.25,-562 39.5,-557 37.75,-562 41.25,-562\"/>\n",
       "</g>\n",
       "<!-- data1_23 -->\n",
       "<g id=\"node55\" class=\"node\">\n",
       "<title>data1_23</title>\n",
       "<text text-anchor=\"middle\" x=\"526.5\" y=\"-489.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">23</text>\n",
       "</g>\n",
       "<!-- data1_23&#45;&gt;data2_HSR -->\n",
       "<g id=\"edge23\" class=\"edge\">\n",
       "<title>data1_23&#45;&gt;data2_HSR</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.96\" d=\"M526.5,-474.89C526.5,-469.57 526.5,-463.63 526.5,-458.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.96\" points=\"528.25,-458 526.5,-453 524.75,-458 528.25,-458\"/>\n",
       "</g>\n",
       "<!-- data1_24 -->\n",
       "<g id=\"node56\" class=\"node\">\n",
       "<title>data1_24</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-177.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">24</text>\n",
       "</g>\n",
       "<!-- data1_24&#45;&gt;data2_SG_8 -->\n",
       "<g id=\"edge24\" class=\"edge\">\n",
       "<title>data1_24&#45;&gt;data2_SG_8</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2.54\" d=\"M434.5,-162.89C434.5,-157.57 434.5,-151.63 434.5,-146.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2.54\" points=\"436.25,-146 434.5,-141 432.75,-146 436.25,-146\"/>\n",
       "</g>\n",
       "<!-- data1_26 -->\n",
       "<g id=\"node57\" class=\"node\">\n",
       "<title>data1_26</title>\n",
       "<text text-anchor=\"middle\" x=\"526.5\" y=\"-593.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">26</text>\n",
       "</g>\n",
       "<!-- data1_26&#45;&gt;data2_CmpR -->\n",
       "<g id=\"edge26\" class=\"edge\">\n",
       "<title>data1_26&#45;&gt;data2_CmpR</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.55\" d=\"M526.5,-578.89C526.5,-573.57 526.5,-567.63 526.5,-562.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.55\" points=\"528.25,-562 526.5,-557 524.75,-562 528.25,-562\"/>\n",
       "</g>\n",
       "<!-- data1_27 -->\n",
       "<g id=\"node58\" class=\"node\">\n",
       "<title>data1_27</title>\n",
       "<text text-anchor=\"middle\" x=\"624.5\" y=\"-386.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">27</text>\n",
       "</g>\n",
       "<!-- data1_27&#45;&gt;data2_OxPh NADPH ** -->\n",
       "<g id=\"edge27\" class=\"edge\">\n",
       "<title>data1_27&#45;&gt;data2_OxPh NADPH **</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.89\" d=\"M624.5,-371.9C624.5,-366.68 624.5,-360.84 624.5,-355.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.89\" points=\"626.25,-355.28 624.5,-350.28 622.75,-355.28 626.25,-355.28\"/>\n",
       "</g>\n",
       "<!-- data1_28 -->\n",
       "<g id=\"node59\" class=\"node\">\n",
       "<title>data1_28</title>\n",
       "<text text-anchor=\"middle\" x=\"206.5\" y=\"-593.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">28</text>\n",
       "</g>\n",
       "<!-- data1_28&#45;&gt;data2_Biofilm&#45;2 -->\n",
       "<g id=\"edge28\" class=\"edge\">\n",
       "<title>data1_28&#45;&gt;data2_Biofilm&#45;2</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2.67\" d=\"M206.5,-578.89C206.5,-573.57 206.5,-567.63 206.5,-562.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2.67\" points=\"208.25,-562 206.5,-557 204.75,-562 208.25,-562\"/>\n",
       "</g>\n",
       "<!-- data1_3 -->\n",
       "<g id=\"node60\" class=\"node\">\n",
       "<title>data1_3</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-489.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">3</text>\n",
       "</g>\n",
       "<!-- data1_3&#45;&gt;data2_Fur&#45;Like -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>data1_3&#45;&gt;data2_Fur&#45;Like</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.21\" d=\"M434.5,-474.89C434.5,-469.57 434.5,-463.63 434.5,-458.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.21\" points=\"436.25,-458 434.5,-453 432.75,-458 436.25,-458\"/>\n",
       "</g>\n",
       "<!-- data1_30 -->\n",
       "<g id=\"node61\" class=\"node\">\n",
       "<title>data1_30</title>\n",
       "<text text-anchor=\"middle\" x=\"100.5\" y=\"-593.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">30</text>\n",
       "</g>\n",
       "<!-- data1_30&#45;&gt;data2_Biofilm&#45;1 -->\n",
       "<g id=\"edge30\" class=\"edge\">\n",
       "<title>data1_30&#45;&gt;data2_Biofilm&#45;1</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"1.93\" d=\"M81.87,-578.89C75.6,-573.14 68.54,-566.66 62.01,-560.66\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"1.93\" points=\"62.89,-559.09 58.02,-557 60.52,-561.67 62.89,-559.09\"/>\n",
       "</g>\n",
       "<!-- data1_30&#45;&gt;data2_SphR -->\n",
       "<g id=\"edge29\" class=\"edge\">\n",
       "<title>data1_30&#45;&gt;data2_SphR</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2.7\" d=\"M102.64,-578.89C103.3,-573.57 104.05,-567.63 104.75,-562.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2.7\" points=\"106.49,-562.18 105.37,-557 103.02,-561.75 106.49,-562.18\"/>\n",
       "</g>\n",
       "<!-- data1_31 -->\n",
       "<g id=\"node62\" class=\"node\">\n",
       "<title>data1_31</title>\n",
       "<text text-anchor=\"middle\" x=\"730.5\" y=\"-178.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">31</text>\n",
       "</g>\n",
       "<!-- data1_31&#45;&gt;data2_Translation -->\n",
       "<g id=\"edge31\" class=\"edge\">\n",
       "<title>data1_31&#45;&gt;data2_Translation</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.96\" d=\"M730.5,-163.9C730.5,-158.68 730.5,-152.84 730.5,-147.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.96\" points=\"732.25,-147.28 730.5,-142.28 728.75,-147.28 732.25,-147.28\"/>\n",
       "</g>\n",
       "<!-- data1_32 -->\n",
       "<g id=\"node63\" class=\"node\">\n",
       "<title>data1_32</title>\n",
       "<text text-anchor=\"middle\" x=\"526.5\" y=\"-282.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">32</text>\n",
       "</g>\n",
       "<!-- data1_32&#45;&gt;data2_Purine Metabolism -->\n",
       "<g id=\"edge32\" class=\"edge\">\n",
       "<title>data1_32&#45;&gt;data2_Purine Metabolism</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.39\" d=\"M526.5,-267.9C526.5,-262.68 526.5,-256.84 526.5,-251.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.39\" points=\"528.25,-251.28 526.5,-246.28 524.75,-251.28 528.25,-251.28\"/>\n",
       "</g>\n",
       "<!-- data1_33 -->\n",
       "<g id=\"node64\" class=\"node\">\n",
       "<title>data1_33</title>\n",
       "<text text-anchor=\"middle\" x=\"526.5\" y=\"-385.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">33</text>\n",
       "</g>\n",
       "<!-- data1_33&#45;&gt;data2_NtcA&#45;4 -->\n",
       "<g id=\"edge33\" class=\"edge\">\n",
       "<title>data1_33&#45;&gt;data2_NtcA&#45;4</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2.57\" d=\"M526.5,-370.89C526.5,-365.57 526.5,-359.63 526.5,-354.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2.57\" points=\"528.25,-354 526.5,-349 524.75,-354 528.25,-354\"/>\n",
       "</g>\n",
       "<!-- data1_34 -->\n",
       "<g id=\"node65\" class=\"node\">\n",
       "<title>data1_34</title>\n",
       "<text text-anchor=\"middle\" x=\"210.5\" y=\"-282.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">34</text>\n",
       "</g>\n",
       "<!-- data1_34&#45;&gt;data2_PSII&#45;Cofactors -->\n",
       "<g id=\"edge35\" class=\"edge\">\n",
       "<title>data1_34&#45;&gt;data2_PSII&#45;Cofactors</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2.42\" d=\"M200.9,-267.9C197.91,-262.57 194.56,-256.6 191.38,-250.94\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2.42\" points=\"192.74,-249.78 188.76,-246.28 189.68,-251.5 192.74,-249.78\"/>\n",
       "</g>\n",
       "<!-- data1_34&#45;&gt;data2_U_5 -->\n",
       "<g id=\"edge34\" class=\"edge\">\n",
       "<title>data1_34&#45;&gt;data2_U_5</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2.44\" d=\"M220.1,-267.9C223.33,-262.15 226.97,-255.66 230.37,-249.61\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2.44\" points=\"231.98,-250.31 232.91,-245.09 228.93,-248.59 231.98,-250.31\"/>\n",
       "</g>\n",
       "<!-- data1_35 -->\n",
       "<g id=\"node66\" class=\"node\">\n",
       "<title>data1_35</title>\n",
       "<text text-anchor=\"middle\" x=\"624.5\" y=\"-593.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">35</text>\n",
       "</g>\n",
       "<!-- data1_35&#45;&gt;data2_Comp&#45;1 -->\n",
       "<g id=\"edge36\" class=\"edge\">\n",
       "<title>data1_35&#45;&gt;data2_Comp&#45;1</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2.93\" d=\"M624.5,-578.89C624.5,-573.57 624.5,-567.63 624.5,-562.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2.93\" points=\"626.25,-562 624.5,-557 622.75,-562 626.25,-562\"/>\n",
       "</g>\n",
       "<!-- data1_36 -->\n",
       "<g id=\"node67\" class=\"node\">\n",
       "<title>data1_36</title>\n",
       "<text text-anchor=\"middle\" x=\"331.5\" y=\"-282.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">36</text>\n",
       "</g>\n",
       "<!-- data1_36&#45;&gt;data2_Phage Related -->\n",
       "<g id=\"edge37\" class=\"edge\">\n",
       "<title>data1_36&#45;&gt;data2_Phage Related</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.32\" d=\"M331.5,-267.9C331.5,-262.68 331.5,-256.84 331.5,-251.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.32\" points=\"333.39,-251.28 331.5,-246.28 329.61,-251.28 333.39,-251.28\"/>\n",
       "</g>\n",
       "<!-- data1_38 -->\n",
       "<g id=\"node68\" class=\"node\">\n",
       "<title>data1_38</title>\n",
       "<text text-anchor=\"middle\" x=\"730.5\" y=\"-593.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">38</text>\n",
       "</g>\n",
       "<!-- data1_38&#45;&gt;data2_Comp&#45;2 -->\n",
       "<g id=\"edge38\" class=\"edge\">\n",
       "<title>data1_38&#45;&gt;data2_Comp&#45;2</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.87\" d=\"M730.5,-578.89C730.5,-573.57 730.5,-567.63 730.5,-562.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.87\" points=\"732.25,-562 730.5,-557 728.75,-562 732.25,-562\"/>\n",
       "</g>\n",
       "<!-- data1_39 -->\n",
       "<g id=\"node69\" class=\"node\">\n",
       "<title>data1_39</title>\n",
       "<text text-anchor=\"middle\" x=\"526.5\" y=\"-177.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">39</text>\n",
       "</g>\n",
       "<!-- data1_39&#45;&gt;data2_STR -->\n",
       "<g id=\"edge39\" class=\"edge\">\n",
       "<title>data1_39&#45;&gt;data2_STR</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2.81\" d=\"M526.5,-162.89C526.5,-157.57 526.5,-151.63 526.5,-146.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2.81\" points=\"528.25,-146 526.5,-141 524.75,-146 528.25,-146\"/>\n",
       "</g>\n",
       "<!-- data1_4 -->\n",
       "<g id=\"node70\" class=\"node\">\n",
       "<title>data1_4</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-281.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">4</text>\n",
       "</g>\n",
       "<!-- data1_4&#45;&gt;data2_PhoTax -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>data1_4&#45;&gt;data2_PhoTax</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.49\" d=\"M434.5,-266.89C434.5,-261.57 434.5,-255.63 434.5,-250.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.49\" points=\"436.25,-250 434.5,-245 432.75,-250 436.25,-250\"/>\n",
       "</g>\n",
       "<!-- data1_40 -->\n",
       "<g id=\"node71\" class=\"node\">\n",
       "<title>data1_40</title>\n",
       "<text text-anchor=\"middle\" x=\"67.5\" y=\"-386.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">40</text>\n",
       "</g>\n",
       "<!-- data1_40&#45;&gt;data2_NblS&#45;RpaB&#45;SrrA&#45;2  -->\n",
       "<g id=\"edge40\" class=\"edge\">\n",
       "<title>data1_40&#45;&gt;data2_NblS&#45;RpaB&#45;SrrA&#45;2 </title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.57\" d=\"M67.5,-371.9C67.5,-366.68 67.5,-360.84 67.5,-355.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.57\" points=\"69.25,-355.28 67.5,-350.28 65.75,-355.28 69.25,-355.28\"/>\n",
       "</g>\n",
       "<!-- data1_42 -->\n",
       "<g id=\"node72\" class=\"node\">\n",
       "<title>data1_42</title>\n",
       "<text text-anchor=\"middle\" x=\"331.5\" y=\"-593.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">42</text>\n",
       "</g>\n",
       "<!-- data1_42&#45;&gt;data2_CMM -->\n",
       "<g id=\"edge41\" class=\"edge\">\n",
       "<title>data1_42&#45;&gt;data2_CMM</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.14\" d=\"M331.5,-578.89C331.5,-573.57 331.5,-567.63 331.5,-562.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.14\" points=\"333.25,-562 331.5,-557 329.75,-562 333.25,-562\"/>\n",
       "</g>\n",
       "<!-- data1_43 -->\n",
       "<g id=\"node73\" class=\"node\">\n",
       "<title>data1_43</title>\n",
       "<text text-anchor=\"middle\" x=\"67.5\" y=\"-281.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">43</text>\n",
       "</g>\n",
       "<!-- data1_43&#45;&gt;data2_PS -->\n",
       "<g id=\"edge42\" class=\"edge\">\n",
       "<title>data1_43&#45;&gt;data2_PS</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.9\" d=\"M67.5,-266.89C67.5,-261.57 67.5,-255.63 67.5,-250.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.9\" points=\"69.25,-250 67.5,-245 65.75,-250 69.25,-250\"/>\n",
       "</g>\n",
       "<!-- data1_5 -->\n",
       "<g id=\"node74\" class=\"node\">\n",
       "<title>data1_5</title>\n",
       "<text text-anchor=\"middle\" x=\"66.5\" y=\"-73.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">5</text>\n",
       "</g>\n",
       "<!-- data1_5&#45;&gt;data2_U_11 -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>data1_5&#45;&gt;data2_U_11</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"1.74\" d=\"M58.25,-58.89C55.64,-53.47 52.71,-47.39 49.96,-41.69\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"1.74\" points=\"51.44,-40.75 47.7,-37 48.29,-42.27 51.44,-40.75\"/>\n",
       "</g>\n",
       "<!-- data1_5&#45;&gt;data2_U_12 -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>data1_5&#45;&gt;data2_U_12</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.06\" d=\"M75.05,-58.89C77.77,-53.47 80.8,-47.39 83.66,-41.69\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.06\" points=\"85.33,-42.26 86,-37 82.2,-40.69 85.33,-42.26\"/>\n",
       "</g>\n",
       "<!-- data1_6 -->\n",
       "<g id=\"node75\" class=\"node\">\n",
       "<title>data1_6</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-594.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">6</text>\n",
       "</g>\n",
       "<!-- data1_6&#45;&gt;data2_Carbon Transport -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>data1_6&#45;&gt;data2_Carbon Transport</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.63\" d=\"M434.5,-579.9C434.5,-574.68 434.5,-568.84 434.5,-563.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.63\" points=\"436.25,-563.28 434.5,-558.28 432.75,-563.28 436.25,-563.28\"/>\n",
       "</g>\n",
       "<!-- data1_7 -->\n",
       "<g id=\"node76\" class=\"node\">\n",
       "<title>data1_7</title>\n",
       "<text text-anchor=\"middle\" x=\"730.5\" y=\"-490.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">7</text>\n",
       "</g>\n",
       "<!-- data1_7&#45;&gt;data2_Membrane Biogenesis -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>data1_7&#45;&gt;data2_Membrane Biogenesis</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.85\" d=\"M730.5,-475.9C730.5,-470.68 730.5,-464.84 730.5,-459.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.85\" points=\"732.25,-459.28 730.5,-454.28 728.75,-459.28 732.25,-459.28\"/>\n",
       "</g>\n",
       "<!-- data1_8 -->\n",
       "<g id=\"node77\" class=\"node\">\n",
       "<title>data1_8</title>\n",
       "<text text-anchor=\"middle\" x=\"624.5\" y=\"-281.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">8</text>\n",
       "</g>\n",
       "<!-- data1_8&#45;&gt;data2_RpaA/RpaB -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>data1_8&#45;&gt;data2_RpaA/RpaB</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.92\" d=\"M624.5,-266.89C624.5,-261.57 624.5,-255.63 624.5,-250.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.92\" points=\"626.25,-250 624.5,-245 622.75,-250 626.25,-250\"/>\n",
       "</g>\n",
       "<!-- data1_9 -->\n",
       "<g id=\"node78\" class=\"node\">\n",
       "<title>data1_9</title>\n",
       "<text text-anchor=\"middle\" x=\"730.5\" y=\"-385.3\" font-family=\"Helvetica,sans-Serif\" font-weight=\"bold\" font-size=\"14.00\" fill=\"red\">9</text>\n",
       "</g>\n",
       "<!-- data1_9&#45;&gt;data2_OxiTol -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>data1_9&#45;&gt;data2_OxiTol</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"2.97\" d=\"M730.5,-370.89C730.5,-365.57 730.5,-359.63 730.5,-354.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"2.97\" points=\"732.25,-354 730.5,-349 728.75,-354 732.25,-354\"/>\n",
       "</g>\n",
       "<!-- data1_SG_1 -->\n",
       "<g id=\"node79\" class=\"node\">\n",
       "<title>data1_SG_1</title>\n",
       "<text text-anchor=\"middle\" x=\"67.5\" y=\"-177.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">SG_1</text>\n",
       "</g>\n",
       "<!-- data1_SG_1&#45;&gt;data2_SG_1 -->\n",
       "<g id=\"edge13\" class=\"edge\">\n",
       "<title>data1_SG_1&#45;&gt;data2_SG_1</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"1.69\" d=\"M67.5,-162.89C67.5,-157.57 67.5,-151.63 67.5,-146.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"1.69\" points=\"69.25,-146 67.5,-141 65.75,-146 69.25,-146\"/>\n",
       "</g>\n",
       "<!-- data1_SG_3 -->\n",
       "<g id=\"node80\" class=\"node\">\n",
       "<title>data1_SG_3</title>\n",
       "<text text-anchor=\"middle\" x=\"206.5\" y=\"-177.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">SG_3</text>\n",
       "</g>\n",
       "<!-- data1_SG_3&#45;&gt;data2_SG_3 -->\n",
       "<g id=\"edge16\" class=\"edge\">\n",
       "<title>data1_SG_3&#45;&gt;data2_SG_3</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"3.45\" d=\"M206.5,-162.89C206.5,-157.57 206.5,-151.63 206.5,-146.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"3.45\" points=\"208.25,-146 206.5,-141 204.75,-146 208.25,-146\"/>\n",
       "</g>\n",
       "<!-- data1_SG_5 -->\n",
       "<g id=\"node81\" class=\"node\">\n",
       "<title>data1_SG_5</title>\n",
       "<text text-anchor=\"middle\" x=\"434.5\" y=\"-74.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">SG_5</text>\n",
       "</g>\n",
       "<!-- data1_SG_5&#45;&gt;data2_rpaA/rpopD6  -->\n",
       "<g id=\"edge25\" class=\"edge\">\n",
       "<title>data1_SG_5&#45;&gt;data2_rpaA/rpopD6 </title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"4.33\" d=\"M434.5,-59.9C434.5,-54.68 434.5,-48.84 434.5,-43.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"4.33\" points=\"436.39,-43.28 434.5,-38.28 432.61,-43.28 436.39,-43.28\"/>\n",
       "</g>\n",
       "<!-- data1_SG_8 -->\n",
       "<g id=\"node82\" class=\"node\">\n",
       "<title>data1_SG_8</title>\n",
       "<text text-anchor=\"middle\" x=\"331.5\" y=\"-177.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"14.00\">SG_8</text>\n",
       "</g>\n",
       "<!-- data1_SG_8&#45;&gt;data2_SG_4 -->\n",
       "<g id=\"edge43\" class=\"edge\">\n",
       "<title>data1_SG_8&#45;&gt;data2_SG_4</title>\n",
       "<path fill=\"none\" stroke=\"black\" stroke-width=\"1.93\" d=\"M331.5,-162.89C331.5,-157.57 331.5,-151.63 331.5,-146.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" stroke-width=\"1.93\" points=\"333.25,-146 331.5,-141 329.75,-146 333.25,-146\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.dot.Digraph at 0x7fa925ffec10>"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dots,links = compare_ica(ica_data.M,old.M,cutoff=0.3,\n",
    "                         ortho_file = mapping)\n",
    "links"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "hits_dict = dict((x[0], x[1]) for x in dots)\n",
    "hits = [x[0]for x in dots]\n",
    "for i in ica_data.imodulon_names:\n",
    "    if i in hits and hits.count(i) == 1:\n",
    "        hit_imod = hits_dict[i]\n",
    "        if 'SG' not in hit_imod:\n",
    "            ica_data.rename_imodulons({i:hit_imod})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save iModulon object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon.util import explained_variance\n",
    "from pymodulon.io import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add iModulon sizes and explained variance\n",
    "for im in ica_data.imodulon_names:\n",
    "    ica_data.imodulon_table.loc[im,'imodulon_size'] = len(ica_data.view_imodulon(im))\n",
    "    ica_data.imodulon_table.loc[im,'explained_variance'] = explained_variance(ica_data,imodulons=im)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This will save your iModulon table, your thresholds, and any other information stored in the ica_data object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'../data/all_strict'"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_to_json(ica_data, path.join(data_dir,'ica_wip.json.gz'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you prefer to view and edit your iModulon table in excel, save it as a CSV and reload the iModulon as before"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data.imodulon_table.to_csv(path.join(data_dir,'imodulon_table_raw.csv'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
